// @thatopen/ui-obc/dist/index.js
import * as x from "@thatopen/components";
import { TechnicalDrawings as at, DxfManager as ht } from "@thatopen/components";
import * as q from "@thatopen/components-front";
import { LitElement as tt, css as et, html as it } from "lit";
import { property as R, state as Ut } from "lit/decorators.js";
import * as I from "three";
import { CSS2DObject as Xt } from "three/examples/jsm/renderers/CSS2DRenderer.js";
import { createRef as z, ref as X } from "lit/directives/ref.js";
import * as c from "@thatopen/ui";
import { Manager as Z } from "@thatopen/ui";
import * as bt from "@thatopen/fragments";
class Yt {
  constructor(t, e) {
    this._group = new I.Group(), this._frustum = new I.Frustum(), this._frustumMat = new I.Matrix4(), this._regenerateDelay = 200, this._regenerateCounter = 0, this.material = new I.LineBasicMaterial({ color: "#2e3338" }), this.numbers = new I.Group(), this.maxRegenerateRetrys = 4, this.gridsFactor = 5, this._scaleX = 1, this._scaleY = 1, this._offsetX = 0, this._offsetY = 0, this._camera = t, this._container = e;
    const i = this.newGrid(-1), n = this.newGrid(-2);
    this.grids = { main: i, secondary: n }, this._group.add(n, i, this.numbers);
  }
  set scaleX(t) {
    this._scaleX = t, this.regenerate();
  }
  get scaleX() {
    return this._scaleX;
  }
  set scaleY(t) {
    this._scaleY = t, this.regenerate();
  }
  get scaleY() {
    return this._scaleY;
  }
  set offsetX(t) {
    this._offsetX = t, this.regenerate();
  }
  get offsetX() {
    return this._offsetX;
  }
  set offsetY(t) {
    this._offsetY = t, this.regenerate();
  }
  get offsetY() {
    return this._offsetY;
  }
  get() {
    return this._group;
  }
  dispose() {
    const { main: t, secondary: e } = this.grids;
    t.removeFromParent(), e.removeFromParent(), t.geometry.dispose(), t.material.dispose(), e.geometry.dispose(), e.material.dispose();
  }
  regenerate() {
    if (!this.isGridReady()) {
      if (this._regenerateCounter++, this._regenerateCounter > this.maxRegenerateRetrys)
        throw new Error("Grid could not be regenerated");
      setTimeout(() => this.regenerate, this._regenerateDelay);
      return;
    }
    this._regenerateCounter = 0, this._camera.updateMatrix(), this._camera.updateMatrixWorld();
    const e = this._frustumMat.multiplyMatrices(
      this._camera.projectionMatrix,
      this._camera.matrixWorldInverse
    );
    this._frustum.setFromProjectionMatrix(e);
    const { planes: i } = this._frustum, n = i[0].constant * -i[0].normal.x, o = i[1].constant * -i[1].normal.x, s = i[2].constant * -i[2].normal.y, a = i[3].constant * -i[3].normal.y, l = Math.abs(n - o), d = Math.abs(a - s), { clientWidth: m, clientHeight: u } = this._container, p = Math.max(m, u), h = Math.max(l, d) / p, f = Math.ceil(Math.log10(l / this.scaleX)), w = Math.ceil(Math.log10(d / this.scaleY)), _ = 10 ** (f - 2) * this.scaleX, g = 10 ** (w - 2) * this.scaleY, y = _ * this.gridsFactor, v = g * this.gridsFactor, $ = Math.ceil(d / v), S = Math.ceil(l / y), O = Math.ceil(d / g), F = Math.ceil(l / _), A = _ * Math.ceil(o / _), B = g * Math.ceil(s / g), L = y * Math.ceil(o / y), U = v * Math.ceil(s / v), j = [...this.numbers.children];
    for (const M of j)
      M.removeFromParent();
    this.numbers.children = [];
    const C = [], k = 9 * h, D = 1e4, K = L + this._offsetX, Ot = Math.round(Math.abs(K / this.scaleX) * D) / D, Lt = (S - 1) * y, zt = Math.round(Math.abs((K + Lt) / this.scaleX) * D) / D, Ft = Math.max(Ot, zt).toString().length * k;
    let nt = Math.ceil(Ft / y) * y;
    for (let M = 0; M < S; M++) {
      let T = L + M * y;
      C.push(T, a, 0, T, s, 0), T = Math.round(T * D) / D, nt = Math.round(nt * D) / D;
      const W = T % nt;
      if (!(y < 1 || v < 1) && Math.abs(W) > 0.01)
        continue;
      const rt = this.newNumber((T + this._offsetX) / this.scaleX), Vt = 12 * h;
      rt.position.set(T, s + Vt, 0);
    }
    for (let M = 0; M < $; M++) {
      const T = U + M * v;
      C.push(o, T, 0, n, T, 0);
      const W = this.newNumber(T / this.scaleY);
      let st = 12;
      W.element.textContent && (st += 4 * W.element.textContent.length);
      const rt = st * h;
      W.position.set(o + rt, T, 0);
    }
    const ot = [];
    for (let M = 0; M < F; M++) {
      const T = A + M * _;
      ot.push(T, a, 0, T, s, 0);
    }
    for (let M = 0; M < O; M++) {
      const T = B + M * g;
      ot.push(o, T, 0, n, T, 0);
    }
    const At = new I.BufferAttribute(new Float32Array(C), 3), Bt = new I.BufferAttribute(new Float32Array(ot), 3), { main: jt, secondary: Nt } = this.grids;
    jt.geometry.setAttribute("position", At), Nt.geometry.setAttribute("position", Bt);
  }
  newNumber(t) {
    const e = document.createElement("bim-label");
    e.textContent = String(Math.round(t * 100) / 100);
    const i = new Xt(e);
    return this.numbers.add(i), i;
  }
  newGrid(t) {
    const e = new I.BufferGeometry(), i = new I.LineSegments(e, this.material);
    return i.frustumCulled = !1, i.renderOrder = t, i;
  }
  isGridReady() {
    const t = this._camera.projectionMatrix.elements;
    for (let e = 0; e < t.length; e++) {
      const i = t[e];
      if (Number.isNaN(i))
        return !1;
    }
    return !0;
  }
}
var Gt = Object.defineProperty, Ht = Object.getOwnPropertyDescriptor, J = (r, t, e, i) => {
  for (var n = Ht(t, e), o = r.length - 1, s; o >= 0; o--)
    (s = r[o]) && (n = s(t, e, n) || n);
  return n && Gt(t, e, n), n;
};
const dt = class dt extends tt {
  constructor() {
    super(...arguments), this._grid = null, this._world = null, this.resize = () => {
      this._world && this._grid && this._grid.regenerate();
    };
  }
  set gridColor(t) {
    if (this._gridColor = t, !(t && this._grid))
      return;
    const e = Number(t.replace("#", "0x"));
    Number.isNaN(e) || this._grid.material.color.setHex(e);
  }
  get gridColor() {
    return this._gridColor;
  }
  set gridScaleX(t) {
    this._gridScaleX = t, t && this._grid && (this._grid.scaleX = t);
  }
  get gridScaleX() {
    return this._gridScaleX;
  }
  set gridScaleY(t) {
    this._gridScaleY = t, t && this._grid && (this._grid.scaleY = t);
  }
  get gridScaleY() {
    return this._gridScaleY;
  }
  get gridOffsetX() {
    var t;
    return ((t = this._grid) == null ? void 0 : t.offsetX) || 0;
  }
  set gridOffsetX(t) {
    this._grid && (this._grid.offsetX = t);
  }
  get gridOffsetY() {
    var t;
    return ((t = this._grid) == null ? void 0 : t.offsetY) || 0;
  }
  set gridOffsetY(t) {
    this._grid && (this._grid.offsetY = t);
  }
  set components(t) {
    this.dispose();
    const i = t.get(x.Worlds).create();
    this._world = i, i.scene = new x.SimpleScene(t), i.scene.setup(), i.renderer = new q.RendererWith2D(t, this);
    const n = new x.OrthoPerspectiveCamera(t);
    i.camera = n;
    const o = new Yt(n.threeOrtho, this);
    this._grid = o, i.scene.three.add(o.get()), n.controls.addEventListener(
      "update",
      () => o.regenerate()
    ), setTimeout(async () => {
      i.camera.updateAspect(), n.set("Plan"), await n.controls.setLookAt(0, 0, 100, 0, 0, 0), await n.projection.set("Orthographic"), n.controls.dollySpeed = 3, n.controls.draggingSmoothTime = 0.085, n.controls.maxZoom = 1e3, n.controls.zoom(4);
    });
  }
  get world() {
    return this._world;
  }
  dispose() {
    var t;
    (t = this.world) == null || t.dispose(), this._world = null, this._grid = null;
  }
  connectedCallback() {
    super.connectedCallback(), new ResizeObserver(this.resize).observe(this);
  }
  disconnectedCallback() {
    super.disconnectedCallback(), this.dispose();
  }
  render() {
    return it`<slot></slot>`;
  }
};
dt.styles = et`
    :host {
      position: relative;
      display: flex;
      min-width: 0px;
      height: 100%;
      background-color: var(--bim-ui_bg-base);
    }
  `;
let N = dt;
J([
  R({ type: String, attribute: "grid-color", reflect: !0 })
], N.prototype, "gridColor");
J([
  R({ type: Number, attribute: "grid-scale-x", reflect: !0 })
], N.prototype, "gridScaleX");
J([
  R({ type: Number, attribute: "grid-scale-y", reflect: !0 })
], N.prototype, "gridScaleY");
J([
  R({ type: Number, attribute: "grid-offset-x", reflect: !0 })
], N.prototype, "gridOffsetX");
J([
  R({ type: Number, attribute: "grid-offset-y", reflect: !0 })
], N.prototype, "gridOffsetY");
var qt = Object.defineProperty, V = (r, t, e, i) => {
  for (var n = void 0, o = r.length - 1, s; o >= 0; o--)
    (s = r[o]) && (n = s(t, e, n) || n);
  return n && qt(t, e, n), n;
};
const mt = class mt extends tt {
  constructor() {
    super(...arguments), this._defaults = {
      size: 60
    }, this._cssMatrix3D = "", this._matrix = new I.Matrix4(), this._onRightClick = new Event("rightclick"), this._onLeftClick = new Event("leftclick"), this._onTopClick = new Event("topclick"), this._onBottomClick = new Event("bottomclick"), this._onFrontClick = new Event("frontclick"), this._onBackClick = new Event("backclick"), this._camera = null, this._epsilon = (t) => Math.abs(t) < 1e-10 ? 0 : t;
  }
  set camera(t) {
    this._camera = t, this.updateOrientation();
  }
  get camera() {
    return this._camera;
  }
  updateOrientation() {
    if (!this.camera)
      return;
    this._matrix.extractRotation(this.camera.matrixWorldInverse);
    const { elements: t } = this._matrix;
    this._cssMatrix3D = `matrix3d(
      ${this._epsilon(t[0])},
      ${this._epsilon(-t[1])},
      ${this._epsilon(t[2])},
      ${this._epsilon(t[3])},
      ${this._epsilon(t[4])},
      ${this._epsilon(-t[5])},
      ${this._epsilon(t[6])},
      ${this._epsilon(t[7])},
      ${this._epsilon(t[8])},
      ${this._epsilon(-t[9])},
      ${this._epsilon(t[10])},
      ${this._epsilon(t[11])},
      ${this._epsilon(t[12])},
      ${this._epsilon(-t[13])},
      ${this._epsilon(t[14])},
      ${this._epsilon(t[15])})
    `;
  }
  render() {
    const t = this.size ?? this._defaults.size;
    return it`
      <style>
        .face,
        .cube {
          width: ${t}px;
          height: ${t}px;
          transform: translateZ(-300px) ${this._cssMatrix3D};
        }

        .face-right {
          translate: ${t / 2}px 0 0;
        }

        .face-left {
          translate: ${-t / 2}px 0 0;
        }

        .face-top {
          translate: 0 ${t / 2}px 0;
        }

        .face-bottom {
          translate: 0 ${-t / 2}px 0;
        }

        .face-front {
          translate: 0 0 ${t / 2}px;
        }

        .face-back {
          translate: 0 0 ${-t / 2}px;
        }
      </style>
      <div class="parent">
        <div class="cube">
          <div
            class="face x-direction face-right"
            @click=${() => this.dispatchEvent(this._onRightClick)}
          >
            ${this.rightText}
          </div>
          <div
            class="face x-direction face-left"
            @click=${() => this.dispatchEvent(this._onLeftClick)}
          >
            ${this.leftText}
          </div>
          <div
            class="face y-direction face-top"
            @click=${() => this.dispatchEvent(this._onTopClick)}
          >
            ${this.topText}
          </div>
          <div
            class="face y-direction face-bottom"
            @click=${() => this.dispatchEvent(this._onBottomClick)}
          >
            ${this.bottomText}
          </div>
          <div
            class="face z-direction face-front"
            @click=${() => this.dispatchEvent(this._onFrontClick)}
          >
            ${this.frontText}
          </div>
          <div
            class="face z-direction face-back"
            @click=${() => this.dispatchEvent(this._onBackClick)}
          >
            ${this.backText}
          </div>
        </div>
      </div>
    `;
  }
};
mt.styles = et`
    :host {
      position: absolute;
      z-index: 999;
      bottom: 1rem;
      right: 1rem;
    }

    .parent {
      perspective: 400px;
    }

    .cube {
      position: relative;
      transform-style: preserve-3d;
    }

    .face {
      position: absolute;
      display: flex;
      justify-content: center;
      user-select: none;
      align-items: center;
      cursor: pointer;
      text-align: center;
      transition: all 0.2s;
      color: var(--bim-view-cube--c, white);
      font-size: var(--bim-view-cube--fz, --bim-ui_size-2xl);
    }

    .x-direction {
      // background-color: var(--bim-view-cube_x--bgc, #c93830DD);
      background-color: var(--bim-view-cube_x--bgc, #01a6bcde);
    }

    .x-direction:hover {
      background-color: var(--bim-ui_accent-base, white);
    }

    .y-direction {
      // background-color: var(--bim-view-cube_y--bgc, #54ff19DD);
      background-color: var(--bim-view-cube_y--bgc, #8d0ec8de);
    }

    .y-direction:hover {
      background-color: var(--bim-ui_accent-base, white);
    }

    .z-direction {
      // background-color: var(--bim-view-cube_z--bgc, #3041c9DD);
      background-color: var(--bim-view-cube_z--bgc, #2718afde);
    }

    .z-direction:hover {
      background-color: var(--bim-ui_accent-base, white);
    }

    .face-front {
      transform: rotateX(180deg);
    }

    .face-back {
      transform: rotateZ(180deg);
    }

    .face-top {
      transform: rotateX(90deg);
    }

    .face-bottom {
      transform: rotateX(270deg);
    }

    .face-right {
      transform: rotateY(-270deg) rotateX(180deg);
    }

    .face-left {
      transform: rotateY(-90deg) rotateX(180deg);
    }
  `;
let P = mt;
V([
  R({ type: Number, reflect: !0 })
], P.prototype, "size");
V([
  R({ type: String, attribute: "right-text", reflect: !0 })
], P.prototype, "rightText");
V([
  R({ type: String, attribute: "left-text", reflect: !0 })
], P.prototype, "leftText");
V([
  R({ type: String, attribute: "top-text", reflect: !0 })
], P.prototype, "topText");
V([
  R({ type: String, attribute: "bottom-text", reflect: !0 })
], P.prototype, "bottomText");
V([
  R({ type: String, attribute: "front-text", reflect: !0 })
], P.prototype, "frontText");
V([
  R({ type: String, attribute: "back-text", reflect: !0 })
], P.prototype, "backText");
V([
  Ut()
], P.prototype, "_cssMatrix3D");
var Wt = Object.defineProperty, Jt = (r, t, e, i) => {
  for (var n = void 0, o = r.length - 1, s; o >= 0; o--)
    (s = r[o]) && (n = s(t, e, n) || n);
  return n && Wt(t, e, n), n;
};
const ut = class ut extends tt {
  constructor() {
    super(...arguments), this.world = null, this._components = null, this._viewport = z();
  }
  set components(t) {
    var e;
    if (this._components = t, this.components) {
      const i = this.components.get(x.Worlds);
      this.world = i.create(), this.world.name = this.name;
    } else
      (e = this.world) == null || e.dispose(), this.world = null;
  }
  get components() {
    return this._components;
  }
  connectedCallback() {
    super.connectedCallback(), this.world && (this.world.enabled = !0);
  }
  disconnectedCallback() {
    super.disconnectedCallback(), this.world && (this.world.enabled = !1);
  }
  dispose() {
    this.components = null, this.remove();
  }
  firstUpdated() {
    const { value: t } = this._viewport;
    if (!(this.components && t && this.world))
      return;
    const e = new x.SimpleScene(this.components);
    this.world.scene = e, e.setup(), e.three.background = null;
    const i = new x.SimpleRenderer(this.components, t);
    this.world.renderer = i;
    const n = new x.OrthoPerspectiveCamera(this.components);
    this.world.camera = n;
    const o = this.components.get(x.Grids).create(this.world);
    o.material.uniforms.uColor.value = new I.Color(4342338), o.material.uniforms.uSize1.value = 2, o.material.uniforms.uSize2.value = 8;
  }
  onSlotChange() {
    const t = new Event("slotchange");
    this.dispatchEvent(t);
  }
  render() {
    return it` <bim-viewport ${X(this._viewport)}>
      <slot @slotchange=${this.onSlotChange}></slot>
    </bim-viewport>`;
  }
};
ut.styles = et``;
let Q = ut;
Jt([
  R({ type: String, reflect: !0 })
], Q.prototype, "name");
const pt = class pt extends tt {
  constructor() {
    super(...arguments), this._canvasRef = z(), this._worldRef = z(), this._overlaysRef = z(), this._toolbarContainerRef = z(), this._vpToolbarRef = z(), this._paperToolbarRef = z(), this._renderer = null, this._resizeObserver = null, this._slots = /* @__PURE__ */ new Map(), this._paperLabels = /* @__PURE__ */ new Map(), this._renderPending = !1, this._pan = { x: 40, y: 40 }, this._zoom = 0.5, this._panning = !1, this._panStart = { x: 0, y: 0 }, this._panOrigin = { x: 0, y: 0 }, this._editingSlot = null, this._selectedSlot = null, this._selectedPaper = null, this._vpSection = null, this._vpNameInput = null, this._vpScaleDropdown = null, this._paperSection = null, this._paperSheetInput = null, this._paperLabelInput = null, this._paperSizeDropdown = null, this._paperOrientDropdown = null, this._dragSlot = null, this._dragType = null, this._dragStartClient = { x: 0, y: 0 }, this._dragTotalMovement = 0, this._dragOrigin = null, this._dragPaper = null, this._dragPaperStartClient = { x: 0, y: 0 }, this._dragPaperOrigin = { x: 0, y: 0 }, this._dragPaperTotalMovement = 0, this.components = null, this._onWheel = (t) => {
      t.preventDefault();
      const e = this.getBoundingClientRect(), i = t.clientX - e.left, n = t.clientY - e.top, o = Math.pow(0.999, t.deltaY), s = Math.max(0.05, Math.min(10, this._zoom * o));
      this._pan.x = i - (i - this._pan.x) * (s / this._zoom), this._pan.y = n - (n - this._pan.y) * (s / this._zoom), this._zoom = s, this._applyTransform(), this.requestRender();
    }, this._onMouseDown = (t) => {
      if (t.button === 0 && !this._editingSlot) {
        const e = t.composedPath().find(
          (i) => i instanceof Element && i.tagName === "BIM-PAPER-SPACE"
        );
        e ? (this._dragPaper = e, this._dragPaperStartClient = { x: t.clientX, y: t.clientY }, this._dragPaperOrigin = { x: e.offsetLeft, y: e.offsetTop }, this._dragPaperTotalMovement = 0) : (this._selectedSlot && this._selectSlot(null), this._selectedPaper && this._selectPaper(null));
      }
      t.button !== 1 && t.button !== 2 || (t.preventDefault(), this._panning = !0, this._panStart = { x: t.clientX, y: t.clientY }, this._panOrigin = { ...this._pan }, this.style.cursor = "grabbing");
    }, this._onMouseMove = (t) => {
      if (this._panning && (this._pan.x = this._panOrigin.x + (t.clientX - this._panStart.x), this._pan.y = this._panOrigin.y + (t.clientY - this._panStart.y), this._applyTransform(), this.requestRender()), this._dragSlot && this._dragType && this._dragOrigin) {
        const e = t.clientX - this._dragStartClient.x, i = t.clientY - this._dragStartClient.y;
        this._dragTotalMovement = Math.max(
          this._dragTotalMovement,
          Math.abs(e) + Math.abs(i)
        ), this._dragType === "move" ? this._applyMoveDrag(e, i) : this._applyResizeDrag(this._dragType, e, i), this.requestRender();
      }
      if (this._dragPaper) {
        const e = t.clientX - this._dragPaperStartClient.x, i = t.clientY - this._dragPaperStartClient.y;
        this._dragPaperTotalMovement = Math.max(
          this._dragPaperTotalMovement,
          Math.abs(e) + Math.abs(i)
        ), this._dragPaper.style.left = `${this._dragPaperOrigin.x + e / this._zoom}px`, this._dragPaper.style.top = `${this._dragPaperOrigin.y + i / this._zoom}px`, this.requestRender();
      }
    }, this._onMouseUp = (t) => {
      if (this._panning && (t.button === 1 || t.button === 2) && (this._panning = !1, this.style.cursor = "default"), this._dragSlot && t.button === 0) {
        const e = this._dragTotalMovement < 4, i = this._dragSlot;
        this._dragSlot = null, this._dragType = null, this._dragOrigin = null, e && this._handleSlotClick(i, t);
      }
      if (this._dragPaper && t.button === 0) {
        const e = this._dragPaperTotalMovement < 4, i = this._dragPaper;
        this._dragPaper = null, e && this._selectPaper(this._selectedPaper === i ? null : i);
      }
    }, this._onContextMenu = (t) => t.preventDefault();
  }
  /**
   * Registers a {@link DrawingViewport} to be rendered inside a
   * `bim-paper-space` element at the given position.
   *
   * @param paper      - The paper sheet that will display this viewport.
   * @param drawingId  - UUID of the {@link TechnicalDrawing} that owns the viewport.
   * @param viewportId - UUID of the {@link DrawingViewport} to render.
   * @param pos        - Top-left position of the viewport in mm, measured from the
   *                     drawing area origin (inside the paper margin).
   */
  addViewport(t, e, i, n) {
    var s;
    const o = {
      drawingId: e,
      viewportId: i,
      paper: t,
      x: n.x,
      y: n.y,
      borderEl: null,
      nameEl: null,
      handles: null
    };
    this._createBorderEl(o), (s = this._overlaysRef.value) == null || s.appendChild(o.borderEl), this._slots.has(t) || (this._slots.set(t, []), this._createPaperLabel(t)), this._slots.get(t).push(o), this.requestRender();
  }
  /**
   * Returns the HTML element that covers the screen area of the given viewport.
   * Useful for registering the viewport as a pointer-event source with an
   * external editor (e.g. `DrawingEditor.registerSource`).
   */
  getViewportElement(t, e) {
    var i;
    return ((i = this._findSlot(t, e)) == null ? void 0 : i.borderEl) ?? null;
  }
  /**
   * Returns the viewport placements registered on the given paper sheet as
   * plain data objects, suitable for passing directly to a DXF exporter.
   *
   * @param paper - The paper sheet to query.
   * @returns Array of `{ vp, x, y }` where `x` and `y` are mm from the
   *          top-left of the drawing area.
   */
  getSlotsForPaper(t) {
    return (this._slots.get(t) ?? []).map((e) => ({ drawingId: e.drawingId, viewportId: e.viewportId, x: e.x, y: e.y }));
  }
  /**
   * Enters edit mode for the given viewport: the board stops intercepting
   * pointer events on that slot so an external editor can take over.
   * The slot stays visually selected but drag/resize/click are suspended.
   */
  enterEditMode(t, e) {
    const i = this._findSlot(t, e);
    if (i) {
      this._editingSlot && this.exitEditMode(), this._editingSlot = i, i.borderEl.style.border = "2px dashed rgba(255,140,0,0.9)", i.borderEl.style.boxShadow = "0 0 0 1px rgba(255,140,0,0.3)", i.borderEl.style.cursor = "crosshair";
      for (const n of Object.values(i.handles))
        n.style.display = "none";
    }
  }
  /**
   * Exits edit mode, returning pointer control to the board.
   */
  exitEditMode() {
    if (!this._editingSlot)
      return;
    const t = this._editingSlot;
    if (this._editingSlot = null, this._selectedSlot === t) {
      t.borderEl.style.border = "2px solid rgba(0,136,255,1)", t.borderEl.style.boxShadow = "0 0 0 1px rgba(0,136,255,0.3)", t.borderEl.style.cursor = "move";
      for (const e of Object.values(t.handles))
        e.style.display = "block";
    } else
      t.borderEl.style.border = "1.5px solid rgba(0,136,255,0.7)", t.borderEl.style.boxShadow = "", t.borderEl.style.cursor = "pointer";
  }
  /**
   * Removes a previously registered viewport from a paper sheet.
   */
  removeViewport(t, e, i) {
    var s;
    const n = this._slots.get(t);
    if (!n)
      return;
    const o = n.findIndex((a) => a.drawingId === e && a.viewportId === i);
    o !== -1 && (n[o] === this._selectedSlot && this._selectSlot(null), n[o].borderEl.remove(), n.splice(o, 1), n.length === 0 && (this._slots.delete(t), (s = this._paperLabels.get(t)) == null || s.remove(), this._paperLabels.delete(t)), this.requestRender());
  }
  /**
   * Schedules a WebGL render pass on the next animation frame.
   * Call this whenever drawing content changes (new annotation, projection
   * update, etc.). Multiple calls within the same frame collapse into one.
   */
  requestRender() {
    this._renderPending || (this._renderPending = !0, requestAnimationFrame(() => {
      this._renderPending = !1, this._doRender();
    }));
  }
  // ── Border / handle / paper-label creation ────────────────────────────────
  _createPaperLabel(t) {
    var i;
    const e = document.createElement("bim-label");
    e.style.cssText = "position:absolute;pointer-events:none;transition:none;--bim-label--c:var(--bim-ui_bg-contrast-100);--bim-label--fz:1.5rem;", (i = this._overlaysRef.value) == null || i.appendChild(e), this._paperLabels.set(t, e);
  }
  _createBorderEl(t) {
    const e = document.createElement("div");
    e.style.cssText = "position:absolute;box-sizing:border-box;border:1.5px solid rgba(0,136,255,0.7);pointer-events:auto;cursor:pointer;";
    const i = [
      { pos: "TL", left: "0%", top: "0%", cursor: "nwse-resize" },
      { pos: "T", left: "50%", top: "0%", cursor: "ns-resize" },
      { pos: "TR", left: "100%", top: "0%", cursor: "nesw-resize" },
      { pos: "L", left: "0%", top: "50%", cursor: "ew-resize" },
      { pos: "R", left: "100%", top: "50%", cursor: "ew-resize" },
      { pos: "BL", left: "0%", top: "100%", cursor: "nesw-resize" },
      { pos: "B", left: "50%", top: "100%", cursor: "ns-resize" },
      { pos: "BR", left: "100%", top: "100%", cursor: "nwse-resize" }
    ], n = {};
    for (const s of i) {
      const a = document.createElement("div");
      a.style.cssText = `position:absolute;width:8px;height:8px;background:#fff;border:1.5px solid rgba(0,136,255,0.9);border-radius:2px;transform:translate(-50%,-50%);display:none;pointer-events:none;left:${s.left};top:${s.top};cursor:${s.cursor};`, a.addEventListener(
        "mousedown",
        (l) => this._onHandleMouseDown(t, s.pos, l)
      ), e.appendChild(a), n[s.pos] = a;
    }
    e.addEventListener(
      "mousedown",
      (s) => this._onBorderMouseDown(t, s)
    );
    const o = document.createElement("bim-label");
    o.textContent = this._nameLabel(this._resolveVp(t)), o.style.cssText = "position:absolute;top:-18px;left:0;pointer-events:none;--bim-label--c:rgba(0,136,255,0.85);", e.appendChild(o), t.borderEl = e, t.nameEl = o, t.handles = n;
  }
  // ── Selection ──────────────────────────────────────────────────────────────
  _selectSlot(t) {
    if (this._selectedSlot) {
      const e = this._selectedSlot;
      e.borderEl.style.border = "1.5px solid rgba(0,136,255,0.7)", e.borderEl.style.boxShadow = "", e.borderEl.style.cursor = "pointer";
      for (const i of Object.values(e.handles))
        i.style.display = "none", i.style.pointerEvents = "none";
      this.dispatchEvent(
        new CustomEvent("viewportdeselect", {
          detail: { paper: e.paper, drawingId: e.drawingId, viewportId: e.viewportId },
          bubbles: !0,
          composed: !0
        })
      );
    }
    if (t && this._selectPaper(null), this._selectedSlot = t, t) {
      t.borderEl.style.border = "2px solid rgba(0,136,255,1)", t.borderEl.style.boxShadow = "0 0 0 1px rgba(0,136,255,0.3)", t.borderEl.style.cursor = "move";
      for (const e of Object.values(t.handles))
        e.style.display = "block", e.style.pointerEvents = "auto";
      this._showVpToolbar(this._resolveVp(t)), this.dispatchEvent(
        new CustomEvent("viewportselect", {
          detail: { paper: t.paper, drawingId: t.drawingId, viewportId: t.viewportId },
          bubbles: !0,
          composed: !0
        })
      );
    } else
      this._hideToolbar();
  }
  _selectPaper(t) {
    if (this._selectedPaper) {
      const e = this._selectedPaper;
      e.style.outline = "", e.style.outlineOffset = "", this._selectedPaper = null, this.dispatchEvent(
        new CustomEvent("paperdeselect", {
          detail: { paper: e },
          bubbles: !0,
          composed: !0
        })
      );
    }
    t && this._selectSlot(null), this._selectedPaper = t, t ? (t.style.outline = "5px solid rgba(255,140,0,0.9)", t.style.outlineOffset = "6px", this._showPaperToolbar(t), this.dispatchEvent(
      new CustomEvent("paperselect", {
        detail: { paper: t },
        bubbles: !0,
        composed: !0
      })
    )) : this._selectedSlot || this._hideToolbar();
  }
  _showVpToolbar(t) {
    this._vpSection && (this._vpSection.label = (t == null ? void 0 : t.name) || "Viewport"), this._vpNameInput && (this._vpNameInput.value = (t == null ? void 0 : t.name) ?? ""), this._vpScaleDropdown && (this._vpScaleDropdown.value = [String((t == null ? void 0 : t.drawingScale) ?? 100)]), this._vpToolbarRef.value && (this._vpToolbarRef.value.style.display = ""), this._paperToolbarRef.value && (this._paperToolbarRef.value.style.display = "none");
    const e = this._toolbarContainerRef.value;
    e && (e.style.display = "block");
  }
  _showPaperToolbar(t) {
    this._paperSection && (this._paperSection.label = t.label || `${t.size} · ${t.orientation}`), this._paperSheetInput && (this._paperSheetInput.value = t.sheetNumber), this._paperLabelInput && (this._paperLabelInput.value = t.label), this._paperSizeDropdown && (this._paperSizeDropdown.value = [t.size]), this._paperOrientDropdown && (this._paperOrientDropdown.value = [t.orientation]), this._vpToolbarRef.value && (this._vpToolbarRef.value.style.display = "none"), this._paperToolbarRef.value && (this._paperToolbarRef.value.style.display = "");
    const e = this._toolbarContainerRef.value;
    e && (e.style.display = "block");
  }
  _hideToolbar() {
    const t = this._toolbarContainerRef.value;
    t && (t.style.display = "none");
  }
  // ── Drag start ────────────────────────────────────────────────────────────
  _onBorderMouseDown(t, e) {
    e.button === 0 && t !== this._editingSlot && (e.stopPropagation(), this._startDrag(t, "move", e));
  }
  _onHandleMouseDown(t, e, i) {
    i.button === 0 && t !== this._editingSlot && (i.stopPropagation(), this._startDrag(t, e, i));
  }
  _startDrag(t, e, i) {
    const n = this._resolveVp(t);
    n && (this._dragSlot = t, this._dragType = e, this._dragStartClient = { x: i.clientX, y: i.clientY }, this._dragTotalMovement = 0, this._dragOrigin = {
      x: t.x,
      y: t.y,
      left: n.left,
      right: n.right,
      top: n.top,
      bottom: n.bottom
    });
  }
  // ── Coordinate helpers ────────────────────────────────────────────────────
  _findSlot(t, e) {
    for (const i of this._slots.values()) {
      const n = i.find((o) => o.drawingId === t && o.viewportId === e);
      if (n)
        return n;
    }
    return null;
  }
  _resolveVp(t) {
    if (!this.components)
      return null;
    const e = this.components.get(at).list.get(t.drawingId);
    return (e == null ? void 0 : e.viewports.get(t.viewportId)) ?? null;
  }
  _getPxPerMm(t) {
    return t.drawingAreaEl.getBoundingClientRect().width / (t.widthMm - 2 * t.margin);
  }
  // ── Drag apply ───────────────────────────────────────────────────────────
  _applyMoveDrag(t, e) {
    const i = this._dragSlot, n = this._dragOrigin, o = this._resolveVp(i);
    if (!o)
      return;
    const s = i.paper, a = this._getPxPerMm(s), d = (s.drawingSlotEl ?? s.drawingAreaEl).getBoundingClientRect(), m = d.width / a, u = d.height / a, p = (o.right - o.left) * (1e3 / o.drawingScale), b = (o.top - o.bottom) * (1e3 / o.drawingScale);
    i.x = Math.max(0, Math.min(m - p, n.x + t / a)), i.y = Math.max(0, Math.min(u - b, n.y + e / a));
  }
  _applyResizeDrag(t, e, i) {
    const n = this._dragSlot, o = this._dragOrigin, s = this._resolveVp(n);
    if (!s)
      return;
    const a = n.paper, l = this._getPxPerMm(a) * (1e3 / s.drawingScale), d = e / l, m = i / l;
    t.includes("L") && (s.left = o.left + d), t.includes("R") && (s.right = o.right + d), t.includes("T") && (s.top = o.top - m), t.includes("B") && (s.bottom = o.bottom - m);
    const u = s.drawingScale, p = this._getPxPerMm(a), h = (a.drawingSlotEl ?? a.drawingAreaEl).getBoundingClientRect(), f = h.width / p, w = h.height / p, _ = (f - n.x) * u / 1e3, g = (w - n.y) * u / 1e3, y = 0.1;
    t.includes("L") && (s.left = Math.max(s.left, s.right - _), s.left = Math.min(s.left, s.right - y)), t.includes("R") && (s.right = Math.min(s.right, s.left + _), s.right = Math.max(s.right, s.left + y)), t.includes("T") && (s.top = Math.min(s.top, s.bottom + g), s.top = Math.max(s.top, s.bottom + y)), t.includes("B") && (s.bottom = Math.max(s.bottom, s.top - g), s.bottom = Math.min(s.bottom, s.top - y));
  }
  // ── Render ────────────────────────────────────────────────────────────────
  _nameLabel(t) {
    if (!t)
      return "";
    const e = t.name ?? "", i = `1:${t.drawingScale}`;
    return e ? `${e} - ${i}` : i;
  }
  /** Traverses the camera's parent chain to find the owning THREE.Scene. */
  _sceneOf(t) {
    let e = t.camera;
    for (; e; ) {
      if (e instanceof I.Scene)
        return e;
      e = e.parent;
    }
    return null;
  }
  _doRender() {
    const t = this._renderer;
    if (!t || this._slots.size === 0)
      return;
    const e = this.getBoundingClientRect(), i = e.width, n = e.height, o = window.devicePixelRatio;
    t.setScissorTest(!1), t.setClearColor(0, 0), t.clear(), t.setScissorTest(!0);
    for (const [s, a] of this._slots) {
      const l = this._paperLabels.get(s);
      if (l) {
        const h = s.getBoundingClientRect(), f = (h.left - e.left - this._pan.x) / this._zoom, w = (h.top - e.top - this._pan.y) / this._zoom - 42;
        l.style.left = `${f}px`, l.style.top = `${w}px`;
        const _ = [s.sheetNumber, s.label].filter(Boolean);
        l.textContent = _.length ? _.join(": ") : "";
      }
      const d = s.drawingAreaEl;
      if (!d)
        continue;
      const m = d.getBoundingClientRect(), u = m.width / (s.widthMm - 2 * s.margin), p = s.drawingSlotEl ?? d, b = p !== d ? p.getBoundingClientRect() : m;
      for (const h of a) {
        const f = this._resolveVp(h), w = f ? this._sceneOf(f) : null, _ = f ? (f.right - f.left) * (1e3 / f.drawingScale) : 0, g = f ? (f.top - f.bottom) * (1e3 / f.drawingScale) : 0, y = b.left - e.left + h.x * u, v = b.top - e.top + h.y * u, $ = _ * u, S = g * u, O = (y - this._pan.x) / this._zoom, F = (v - this._pan.y) / this._zoom, A = $ / this._zoom, B = S / this._zoom;
        if (h.borderEl.style.left = `${O}px`, h.borderEl.style.top = `${F}px`, h.borderEl.style.width = `${A}px`, h.borderEl.style.height = `${B}px`, h.nameEl.textContent = this._nameLabel(f), !w || !f || y + $ <= 0 || y >= i || v + S <= 0 || v >= n)
          continue;
        const L = Math.round(y * o), U = Math.round((n - v - S) * o), j = Math.round($ * o), C = Math.round(S * o);
        t.setScissor(L, U, j, C), t.setViewport(L, U, j, C), t.setClearColor(16777215, 1), t.clear(), t.render(w, f.camera);
      }
    }
    t.setScissorTest(!1);
  }
  // ── Pan / zoom ────────────────────────────────────────────────────────────
  _applyTransform() {
    const t = `translate(${this._pan.x}px,${this._pan.y}px) scale(${this._zoom})`;
    this._worldRef.value && (this._worldRef.value.style.transform = t), this._overlaysRef.value && (this._overlaysRef.value.style.transform = t);
  }
  _handleSlotClick(t, e) {
    e.detail >= 2 ? this.dispatchEvent(
      new CustomEvent("viewportactivate", {
        detail: { paper: t.paper, drawingId: t.drawingId, viewportId: t.viewportId },
        bubbles: !0,
        composed: !0
      })
    ) : this._selectSlot(this._selectedSlot === t ? null : t);
  }
  // ── Lifecycle ─────────────────────────────────────────────────────────────
  connectedCallback() {
    super.connectedCallback(), this.addEventListener("wheel", this._onWheel, { passive: !1 }), this.addEventListener("mousedown", this._onMouseDown), this.addEventListener("contextmenu", this._onContextMenu), window.addEventListener("mousemove", this._onMouseMove), window.addEventListener("mouseup", this._onMouseUp);
  }
  disconnectedCallback() {
    var t, e;
    super.disconnectedCallback(), this.removeEventListener("wheel", this._onWheel), this.removeEventListener("mousedown", this._onMouseDown), this.removeEventListener("contextmenu", this._onContextMenu), window.removeEventListener("mousemove", this._onMouseMove), window.removeEventListener("mouseup", this._onMouseUp), (t = this._resizeObserver) == null || t.disconnect(), (e = this._renderer) == null || e.dispose(), this._renderer = null;
  }
  _buildVpToolbar() {
    const t = this._vpToolbarRef.value;
    if (!t)
      return;
    const e = document.createElement("bim-dropdown");
    e.label = "Scale", e.vertical = !0, e.style.width = "140px";
    for (const a of [25, 50, 100, 200, 500]) {
      const l = document.createElement("bim-option");
      l.label = `1:${a}`, l.value = String(a), e.appendChild(l);
    }
    e.addEventListener("change", () => {
      if (!this._selectedSlot)
        return;
      const a = this._resolveVp(this._selectedSlot);
      if (!a)
        return;
      const l = Number(e.value[0]);
      Number.isNaN(l) || (a.drawingScale = l, this.requestRender());
    });
    const i = document.createElement("bim-text-input");
    i.label = "Name", i.vertical = !0, i.placeholder = "Viewport name...", i.style.width = "140px", i.addEventListener("input", () => {
      if (!this._selectedSlot)
        return;
      const a = this._resolveVp(this._selectedSlot);
      a && (a.name = i.value, this._vpSection && (this._vpSection.label = i.value || "Viewport"), this.requestRender());
    });
    const n = document.createElement("bim-button");
    n.label = "Export DXF", n.icon = "mingcute:file-export-line", n.vertical = !0, n.addEventListener("click", () => {
      if (!this._selectedSlot || !this.components)
        return;
      const a = this._resolveVp(this._selectedSlot);
      if (!a)
        return;
      const d = this.components.get(at).list.get(this._selectedSlot.drawingId);
      if (!d)
        return;
      const m = this.components.get(ht).exporter.export([
        { drawing: d, viewports: [{ viewport: a }] }
      ]);
      this.dispatchEvent(new CustomEvent("viewportdxfexport", {
        detail: { drawingId: this._selectedSlot.drawingId, viewportId: this._selectedSlot.viewportId, dxf: m },
        bubbles: !0,
        composed: !0
      }));
    });
    const o = document.createElement("bim-toolbar-section");
    o.appendChild(i), o.appendChild(e), t.appendChild(o);
    const s = document.createElement("bim-toolbar-section");
    s.label = "Export", s.appendChild(n), t.appendChild(s), this._vpSection = o, this._vpNameInput = i, this._vpScaleDropdown = e;
  }
  _buildPaperToolbar() {
    const t = this._paperToolbarRef.value;
    if (!t)
      return;
    const e = document.createElement("bim-text-input");
    e.label = "Sheet", e.vertical = !0, e.placeholder = "A-01...", e.style.width = "100px", e.addEventListener("input", () => {
      this._selectedPaper && (this._selectedPaper.sheetNumber = e.value, this.requestRender());
    });
    const i = document.createElement("bim-text-input");
    i.label = "Label", i.vertical = !0, i.placeholder = "Sheet label...", i.style.width = "160px", i.addEventListener("input", () => {
      this._selectedPaper && (this._selectedPaper.label = i.value, this._paperSection && (this._paperSection.label = i.value || `${this._selectedPaper.size} · ${this._selectedPaper.orientation}`), this.requestRender());
    });
    const n = document.createElement("bim-dropdown");
    n.label = "Size", n.vertical = !0;
    for (const d of ["A0", "A1", "A2", "A3", "A4"]) {
      const m = document.createElement("bim-option");
      m.label = d, m.value = d, n.appendChild(m);
    }
    n.addEventListener("change", () => {
      this._selectedPaper && (this._selectedPaper.size = n.value[0], this.requestRender());
    });
    const o = document.createElement("bim-dropdown");
    o.label = "Orientation", o.vertical = !0;
    for (const [d, m] of [["Landscape", "landscape"], ["Portrait", "portrait"]]) {
      const u = document.createElement("bim-option");
      u.label = d, u.value = m, o.appendChild(u);
    }
    o.addEventListener("change", () => {
      this._selectedPaper && (this._selectedPaper.orientation = o.value[0], this.requestRender());
    });
    const s = document.createElement("bim-button");
    s.label = "Export DXF", s.icon = "mingcute:file-export-line", s.vertical = !0, s.addEventListener("click", () => {
      var p;
      if (!this._selectedPaper || !this.components)
        return;
      const d = this.components.get(at), m = /* @__PURE__ */ new Map();
      for (const { drawingId: b, viewportId: h, x: f, y: w } of this.getSlotsForPaper(this._selectedPaper)) {
        if (!m.has(b)) {
          const g = d.list.get(b);
          if (!g)
            continue;
          m.set(b, { drawing: g, viewports: [] });
        }
        const _ = (p = d.list.get(b)) == null ? void 0 : p.viewports.get(h);
        _ && m.get(b).viewports.push({ viewport: _, x: f, y: w });
      }
      const u = this.components.get(ht).exporter.export(
        [...m.values()],
        {
          widthMm: this._selectedPaper.widthMm,
          heightMm: this._selectedPaper.heightMm,
          margin: this._selectedPaper.margin
        }
      );
      this.dispatchEvent(new CustomEvent("paperdxfexport", {
        detail: { paper: this._selectedPaper, dxf: u },
        bubbles: !0,
        composed: !0
      }));
    });
    const a = document.createElement("bim-toolbar-section");
    a.appendChild(e), a.appendChild(i), a.appendChild(n), a.appendChild(o), t.appendChild(a);
    const l = document.createElement("bim-toolbar-section");
    l.label = "Export", l.appendChild(s), t.appendChild(l), this._paperSection = a, this._paperSheetInput = e, this._paperLabelInput = i, this._paperSizeDropdown = n, this._paperOrientDropdown = o;
  }
  firstUpdated() {
    var t;
    (t = this._toolbarContainerRef.value) == null || t.addEventListener(
      "mousedown",
      (e) => e.stopPropagation()
    ), this._buildVpToolbar(), this._buildPaperToolbar(), this._applyTransform(), this._renderer = new I.WebGLRenderer({
      canvas: this._canvasRef.value,
      antialias: !0,
      alpha: !0
    }), this._renderer.setPixelRatio(window.devicePixelRatio), this._renderer.autoClear = !1, this._renderer.setSize(this.clientWidth, this.clientHeight), this._resizeObserver = new ResizeObserver(() => {
      var e;
      (e = this._renderer) == null || e.setSize(this.clientWidth, this.clientHeight), this.requestRender();
    }), this._resizeObserver.observe(this);
  }
  render() {
    return it`
      <div class="world" ${X(this._worldRef)}>
        <slot></slot>
      </div>
      <canvas ${X(this._canvasRef)}></canvas>
      <div class="overlays" ${X(this._overlaysRef)}></div>
      <div class="toolbar-container" ${X(this._toolbarContainerRef)}>
        <bim-toolbar style="border: 1px solid var(--bim-ui_bg-contrast-20); box-shadow: 0 2px 8px rgba(0,0,0,0.25);" ${X(this._vpToolbarRef)}></bim-toolbar>
        <bim-toolbar style="border: 1px solid var(--bim-ui_bg-contrast-20); box-shadow: 0 2px 8px rgba(0,0,0,0.25); display: none;" ${X(this._paperToolbarRef)}></bim-toolbar>
      </div>
    `;
  }
};
pt.styles = et`
    :host {
      display: block;
      position: relative;
      overflow: hidden;
      background-color: var(--bim-ui_bg-base);
      background-image:
        linear-gradient(rgba(255,255,255,0.03) 1px, transparent 1px),
        linear-gradient(90deg, rgba(255,255,255,0.03) 1px, transparent 1px);
      background-size: 24px 24px;
      cursor: default;
      user-select: none;
    }

    canvas {
      position: absolute;
      top: 0;
      left: 0;
      width: 100%;
      height: 100%;
      pointer-events: none;
    }

    .world,
    .overlays {
      position: absolute;
      top: 0;
      left: 0;
      transform-origin: 0 0;
    }

    .overlays {
      pointer-events: none;
    }

    .toolbar-container {
      position: absolute;
      bottom: 1rem;
      left: 50%;
      transform: translateX(-50%);
      display: none;
      z-index: 10;
    }
  `;
let lt = pt;
class ci {
  /**
   * Initializes the custom elements for the BIM application.
   *
   * @remarks
   * This method should be called once during the application's initialization.
   *
   */
  static init() {
    Z.defineCustomElement("bim-view-cube", P), Z.defineCustomElement("bim-world-2d", N), Z.defineCustomElement("bim-world", Q), Z.defineCustomElement("bim-sheet-board", lt);
  }
}
const G = (r, t) => {
  const e = t[r], i = (e == null ? void 0 : e.name) ?? r, n = i.trim().split(/\s+/);
  let o, s;
  return n[0] && n[0][0] && (o = n[0][0].toUpperCase(), n[0][1] && (s = n[0][1].toUpperCase())), n[1] && n[1][0] && (s = n[1][0].toUpperCase()), c.html`
    <div style="display: flex; gap: 0.25rem; overflow: hidden;">
      ${!(e != null && e.picture) && (o || s) ? c.html`
        <bim-label
          style=${c.styleMap({
    borderRadius: "999px",
    padding: "0.375rem",
    backgroundColor: "var(--bim-ui_bg-contrast-20)",
    aspectRatio: "1",
    fontSize: "0.7rem"
  })}>${o}${s}</bim-label>
        ` : null}
      <bim-label .img=${e == null ? void 0 : e.picture}>${i}</bim-label>
    </div>
  `;
}, E = {
  users: {
    "jhon.doe@example.com": {
      name: "Jhon Doe"
    }
  },
  priorities: {
    "On hold": {
      icon: "flowbite:circle-pause-outline",
      style: {
        backgroundColor: "var(--bim-ui_bg-contrast-20)",
        "--bim-icon--c": "#767676"
      }
    },
    Minor: {
      icon: "mingcute:arrows-down-fill",
      style: {
        backgroundColor: "var(--bim-ui_bg-contrast-20)",
        "--bim-icon--c": "#4CAF50"
      }
    },
    Normal: {
      icon: "fa6-solid:grip-lines",
      style: {
        backgroundColor: "var(--bim-ui_bg-contrast-20)",
        "--bim-icon--c": "#FB8C00"
      }
    },
    Major: {
      icon: "mingcute:arrows-up-fill",
      style: {
        backgroundColor: "var(--bim-ui_bg-contrast-20)",
        "--bim-icon--c": "#FF5252"
      }
    },
    Critical: {
      icon: "ph:warning",
      style: {
        backgroundColor: "var(--bim-ui_bg-contrast-20)",
        "--bim-icon--c": "#FB8C00"
      }
    }
  },
  statuses: {
    Active: {
      icon: "prime:circle-fill",
      style: {
        backgroundColor: "var(--bim-ui_bg-contrast-20)"
      }
    },
    "In Progress": {
      icon: "prime:circle-fill",
      style: {
        backgroundColor: "#fa89004d",
        "--bim-label--c": "#FB8C00",
        "--bim-icon--c": "#FB8C00"
      }
    },
    "In Review": {
      icon: "prime:circle-fill",
      style: {
        backgroundColor: "#9c6bff4d",
        "--bim-label--c": "#9D6BFF",
        "--bim-icon--c": "#9D6BFF"
      }
    },
    Done: {
      icon: "prime:circle-fill",
      style: {
        backgroundColor: "#4CAF504D",
        "--bim-label--c": "#4CAF50",
        "--bim-icon--c": "#4CAF50"
      }
    },
    Closed: {
      icon: "prime:circle-fill",
      style: {
        backgroundColor: "#414141",
        "--bim-label--c": "#727272",
        "--bim-icon--c": "#727272"
      }
    }
  },
  types: {
    Clash: {
      icon: "gg:close-r",
      style: {
        backgroundColor: "var(--bim-ui_bg-contrast-20)",
        "--bim-icon--c": "#FB8C00"
      }
    },
    Issue: {
      icon: "mdi:bug-outline",
      style: {
        backgroundColor: "var(--bim-ui_bg-contrast-20)",
        "--bim-icon--c": "#FF5252"
      }
    },
    Failure: {
      icon: "mdi:bug-outline",
      style: {
        backgroundColor: "var(--bim-ui_bg-contrast-20)",
        "--bim-icon--c": "#FF5252"
      }
    },
    Inquiry: {
      icon: "majesticons:comment-line",
      style: {
        backgroundColor: "var(--bim-ui_bg-contrast-20)",
        "--bim-icon--c": "#FF5252"
      }
    },
    Fault: {
      icon: "ph:warning",
      style: {
        backgroundColor: "var(--bim-ui_bg-contrast-20)",
        "--bim-icon--c": "#FF5252"
      }
    },
    Remark: {
      icon: "ph:note-blank-bold",
      style: {
        backgroundColor: "var(--bim-ui_bg-contrast-20)",
        "--bim-icon--c": "#FB8C00"
      }
    },
    Request: {
      icon: "mynaui:edit-one",
      style: {
        backgroundColor: "var(--bim-ui_bg-contrast-20)",
        "--bim-icon--c": "#9D6BFF"
      }
    }
  }
}, H = {
  padding: "0.25rem 0.5rem",
  borderRadius: "999px",
  "--bim-label--c": "var(--bim-ui_bg-contrast-100)"
}, Kt = {
  dueDate: (r) => {
    if (typeof r == "string" && r.trim() !== "")
      return new Date(r);
  },
  status: (r) => {
    if (Array.isArray(r) && r.length !== 0)
      return r[0];
  },
  type: (r) => {
    if (Array.isArray(r) && r.length !== 0)
      return r[0];
  },
  priority: (r) => {
    if (Array.isArray(r) && r.length !== 0)
      return r[0];
  },
  stage: (r) => {
    if (Array.isArray(r) && r.length !== 0)
      return r[0];
  },
  assignedTo: (r) => {
    if (Array.isArray(r) && r.length !== 0)
      return r[0];
  },
  labels: (r) => {
    if (Array.isArray(r))
      return new Set(r);
  }
}, gt = (r) => {
  const {
    components: t,
    topic: e,
    value: i,
    onCancel: n,
    onSubmit: o,
    styles: s
  } = r, a = o ?? (() => {
  }), l = t.get(x.BCFTopics), d = (i == null ? void 0 : i.title) ?? (e == null ? void 0 : e.title) ?? x.Topic.default.title, m = (i == null ? void 0 : i.status) ?? (e == null ? void 0 : e.status) ?? x.Topic.default.status, u = (i == null ? void 0 : i.type) ?? (e == null ? void 0 : e.type) ?? x.Topic.default.type, p = (i == null ? void 0 : i.priority) ?? (e == null ? void 0 : e.priority) ?? x.Topic.default.priority, b = (i == null ? void 0 : i.assignedTo) ?? (e == null ? void 0 : e.assignedTo) ?? x.Topic.default.assignedTo, h = (i == null ? void 0 : i.labels) ?? (e == null ? void 0 : e.labels) ?? x.Topic.default.labels, f = (i == null ? void 0 : i.stage) ?? (e == null ? void 0 : e.stage) ?? x.Topic.default.stage, w = (i == null ? void 0 : i.description) ?? (e == null ? void 0 : e.description) ?? x.Topic.default.description, _ = e != null && e.dueDate ? e.dueDate.toISOString().split("T")[0] : null, g = /* @__PURE__ */ new Set([...l.config.statuses]);
  m && g.add(m);
  const y = /* @__PURE__ */ new Set([...l.config.types]);
  u && y.add(u);
  const v = /* @__PURE__ */ new Set([...l.config.priorities]);
  p && v.add(p);
  const $ = /* @__PURE__ */ new Set([...l.config.users]);
  b && $.add(b);
  const S = /* @__PURE__ */ new Set([...l.config.labels]);
  if (h)
    for (const C of h)
      S.add(C);
  const O = /* @__PURE__ */ new Set([...l.config.stages]);
  f && O.add(f);
  const F = z(), A = async () => {
    const { value: C } = F;
    if (!C)
      return;
    const k = c.getElementValue(
      C,
      Kt
    );
    if (e)
      e.set(k), await a(e);
    else {
      const D = l.create(k);
      await a(D);
    }
  }, B = z(), L = (C) => {
    const { value: k } = B;
    if (!k)
      return;
    const D = C.target;
    k.disabled = D.value.trim() === "";
  }, U = `btn-${c.Manager.newRandomId()}`, j = `btn-${c.Manager.newRandomId()}`;
  return c.html`
    <div ${c.ref(F)} style="display: flex; flex-direction: column; gap: 0.75rem;">
      <div style="display: flex; gap: 0.375rem">
        <bim-text-input @input=${L} vertical label="Title" name="title" .value=${d}></bim-text-input>
        ${e ? c.html`
            <bim-dropdown vertical label="Status" name="status" required>
              ${[...g].map((C) => c.html`<bim-option label=${C} .checked=${m === C}></bim-option>`)}
            </bim-dropdown>` : c.html``}
      </div>
      <div style="display: flex; gap: 0.375rem">
        <bim-dropdown vertical label="Type" name="type" required>
          ${[...y].map((C) => c.html`<bim-option label=${C} .checked=${u === C}></bim-option>`)}
        </bim-dropdown>
        <bim-dropdown vertical label="Priority" name="priority">
          ${[...v].map((C) => c.html`<bim-option label=${C} .checked=${p === C}></bim-option>`)}
        </bim-dropdown>
      </div>
      <div style="display: flex; gap: 0.375rem">
        <bim-dropdown vertical label="Labels" name="labels" multiple>
          ${[...S].map((C) => c.html`<bim-option label=${C} .checked=${h ? [...h].includes(C) : !1}></bim-option>`)}
        </bim-dropdown>
        <bim-dropdown vertical label="Assignee" name="assignedTo">
          ${[...$].map((C) => {
    const k = s != null && s.users ? s.users[C] : null, D = k ? k.name : C, K = k == null ? void 0 : k.picture;
    return c.html`<bim-option label=${D} value=${C} .img=${K} .checked=${b === C}></bim-option>`;
  })}
        </bim-dropdown>
      </div>
      <div style="display: flex; gap: 0.375rem">
        <bim-text-input vertical type="date" label="Due Date" name="dueDate" .value=${_}></bim-text-input> 
        <bim-dropdown vertical label="Stage" name="stage">
          ${[...O].map((C) => c.html`<bim-option label=${C} .checked=${f === C}></bim-option>`)}
        </bim-dropdown>
      </div>
      <bim-text-input vertical label="Description" name="description" type="area" .value=${w ?? null}></bim-text-input>
      <div style="justify-content: right; display: flex; gap: 0.375rem">
        <style>
          #${j} {
            background-color: transparent;
          }

          #${j}:hover {
            --bim-label--c: #FF5252;
          }

          #${U}:hover {
            background-color: #329936;
          }
        </style>
        <bim-button id=${j} style="flex: 0" @click=${n} label="Cancel"></bim-button>
        <bim-button id=${U} style="flex: 0" @click=${A} ${c.ref(B)} label=${e ? "Update Topic" : "Add Topic"} icon=${e ? "tabler:refresh" : "mi:add"}></bim-button>
      </div>
    </div>
  `;
}, Zt = (r) => {
  const { components: t, modelUserData: e, worldName: i } = r, n = () => {
    if (!(t && i))
      return;
    const s = [...t.get(x.Worlds).list.values()].find((l) => "name" in l && l.name === i);
    if (!s)
      return;
    const a = document.createElement("input");
    a.type = "file", a.accept = ".ifc", a.onchange = async () => {
      if (a.files === null || a.files.length === 0)
        return;
      const l = a.files[0], d = await l.arrayBuffer(), m = new Uint8Array(d), u = l.name.replace(".ifc", ""), p = t.get(x.FragmentsManager), b = t.get(x.IfcLoader);
      b.settings.autoSetWasm = !1, b.settings.wasm = {
        path: "https://unpkg.com/web-ifc@0.0.77/",
        absolute: !1
      };
      const h = await b.load(m, !0, u, {
        userData: e
      });
      s.scene.three.add(h.object), h.useCamera(s.camera.three), p.core.update(!0);
    }, a.click();
  };
  return c.html`
    <bim-button
      data-ui-id="import-ifc"
      label="Load IFC"
      icon="mage:box-3d-fill"
      @click=${n}
    ></bim-button>
  `;
}, Qt = (r) => c.Component.create(
  Zt,
  r
), te = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  loadIfc: Qt
}, Symbol.toStringTag, { value: "Module" })), ee = (r) => {
  const { components: t, world: e } = r, i = () => {
    const n = document.createElement("input");
    n.type = "file", n.accept = ".frag", n.onchange = async () => {
      if (n.files === null || n.files.length === 0)
        return;
      const o = n.files[0], s = await o.arrayBuffer(), a = new Uint8Array(s), l = o.name.replace(".frag", ""), d = t.get(x.FragmentsManager), m = await d.core.load(a, {
        modelId: l
      });
      e && (e.scene.three.add(m.object), m.useCamera(e.camera.three), d.core.update(!0));
    }, n.click();
  };
  return c.html`
    <bim-button @click=${i}></bim-button>
  `;
}, ie = (r) => {
  const t = c.Component.create(
    ee,
    r
  ), [e] = t;
  return e.label = "Load FRAG", e.icon = "mage:box-3d-fill", t;
}, ne = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  loadFrag: ie
}, Symbol.toStringTag, { value: "Module" })), di = {
  ...te,
  ...ne
}, ct = async (r, t) => {
  const { localId: e, category: i, children: n } = t;
  if (i && n) {
    const o = {
      data: {
        Name: i,
        modelId: r.modelId,
        children: JSON.stringify(n.map((s) => s.localId))
      }
    };
    for (const s of n) {
      const a = await ct(r, s);
      a && (o.children || (o.children = []), o.children.push(a));
    }
    return o;
  }
  if (e !== null) {
    const s = await r.getItem(e).getAttributes();
    if (!s)
      return null;
    const a = {
      data: {
        Name: String(s.getValue("Name")),
        modelId: r.modelId,
        localId: e
      }
    };
    for (const l of n ?? []) {
      const d = await ct(r, l);
      d && (a.children || (a.children = []), a.children.push(d));
    }
    return a;
  }
  return null;
}, oe = async (r) => {
  const t = [];
  for (const e of r) {
    const i = await e.getSpatialStructure(), n = await ct(e, i);
    if (!n)
      continue;
    const o = {
      data: {
        Name: e.modelId,
        modelId: e.modelId
      },
      children: [n]
    };
    t.push(o);
  }
  return t;
}, yt = (r) => {
  const { components: t, models: e } = r, i = r.selectHighlighterName ?? "select", n = ({
    detail: a
  }) => {
    const { cell: l } = a;
    l.column === "Name" && !l.rowData.Name && (l.style.gridColumn = "1 / -1");
  }, o = (a) => {
    a.stopImmediatePropagation();
    const { row: l } = a.detail, d = t.get(q.Highlighter), m = t.get(x.FragmentsManager);
    l.onclick = async () => {
      if (!i)
        return;
      const {
        data: { modelId: u, localId: p, children: b }
      } = l;
      if (!(u && (p !== void 0 || b)))
        return;
      const h = m.list.get(u);
      if (h) {
        if (p !== void 0) {
          const f = await h.getItemsChildren([p]), w = {
            [u]: f.length !== 0 ? new Set(f) : /* @__PURE__ */ new Set([p])
          };
          d.highlightByID(
            i,
            w,
            !0,
            !0
          );
        } else if (b) {
          const f = JSON.parse(b), w = await h.getItemsChildren(f), _ = {
            [u]: w.length !== 0 ? w : f
          };
          d.highlightByID(
            i,
            _,
            !0,
            !0
          );
        }
      }
    };
  }, s = async (a) => {
    if (!a)
      return;
    const l = a;
    l.loadFunction = async () => new Promise((d) => {
      setTimeout(() => {
        d(oe(e));
      });
    }), l.loadData(!0);
  };
  return c.html`
    <bim-table @rowcreated=${o} @cellcreated=${n} ${c.ref(s)} headers-hidden>
      <bim-label slot="missing-data" style="--bim-icon--c: gold" icon="ic:round-warning">
        No models available to display the spatial structure!
      </bim-label>
    </bim-table>
  `;
}, se = (r, t = !0) => {
  const e = c.Component.create(yt, r), [i, n] = e;
  if (i.hiddenColumns = ["modelId", "localId", "children"], i.columns = ["Name"], i.headersHidden = !0, t) {
    const { components: o } = r, s = o.get(x.FragmentsManager);
    s.list.onItemSet.add(
      () => n({ models: s.list.values() })
    ), s.list.onItemDeleted.add(() => n());
  }
  return e;
}, re = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  spatialTree: se,
  spatialTreeTemplate: yt
}, Symbol.toStringTag, { value: "Module" }));
let Y = {};
const ft = {
  _category: "Category",
  _localId: "LocalId",
  _guid: "Guid"
}, ae = (r, t, e, i, n, o) => {
  const s = {
    data: {
      type: "attribute",
      modelId: i,
      localId: n,
      Name: t in ft ? ft[t] : t,
      Value: e,
      dataType: o
    }
  };
  r.children || (r.children = []), r.children.push(s);
}, _t = (r, t, e) => {
  var d;
  r in Y || (Y[r] = /* @__PURE__ */ new Map());
  const i = Y[r], n = t._localId.value, o = (d = t[e.defaultItemNameKey]) == null ? void 0 : d.value, s = t._category.value, a = (o == null ? void 0 : o.toString().length) > 0 ? o.toString() : s ?? String(n);
  if (i.has(n))
    return {
      data: {
        modelId: r,
        localId: n,
        type: "item",
        Name: a
      }
    };
  const l = {
    data: {
      modelId: r,
      localId: n,
      type: "item",
      Name: a
    }
  };
  i.set(n, l);
  for (const m in t) {
    const u = t[m];
    if (!Array.isArray(u))
      ae(l, m, u.value, r, n, u.type);
    else {
      const p = {
        data: {
          Name: m,
          type: "relation"
        }
      };
      l.children || (l.children = []), l.children.push(p);
      for (const b of u) {
        const h = _t(r, b, e);
        p.children || (p.children = []), p.children.push(h);
      }
    }
  }
  return l;
}, le = async (r, t, e) => {
  const i = r.get(x.FragmentsManager);
  Object.keys(t).length === 0 && (Y = {});
  const n = [];
  for (const o in t) {
    const s = i.list.get(o);
    if (!s)
      continue;
    o in Y || (Y[o] = /* @__PURE__ */ new Map());
    const a = Y[o], l = t[o];
    for (const d of l) {
      let m = a.get(d);
      if (m) {
        n.push(m);
        continue;
      }
      const [u] = await s.getItemsData(
        [d],
        e.itemsDataConfig
      );
      m = _t(o, u, e), n.push(m);
    }
  }
  return n;
}, vt = (r) => {
  const t = {
    defaultItemNameKey: "Name",
    itemsDataConfig: {
      attributesDefault: !0,
      relationsDefault: { attributes: !1, relations: !1 },
      relations: {
        IsDefinedBy: { attributes: !0, relations: !0 },
        DefinesOccurrence: { attributes: !1, relations: !1 },
        ContainedInStructure: { attributes: !0, relations: !0 },
        ContainsElements: { attributes: !1, relations: !1 },
        Decomposes: { attributes: !1, relations: !1 },
        ObjectTypeOf: { attributes: !1, relations: !1 }
      }
    },
    ...r
  }, { components: e, modelIdMap: i, emptySelectionWarning: n } = r, o = Object.keys(i).reduce((l, d) => (d.includes("DELTA") || (l[d] = i[d]), l), {}), s = async (l) => {
    if (!l)
      return;
    const d = l;
    d.loadFunction = async () => le(e, o, t), await d.loadData(!0) && d.dispatchEvent(new Event("datacomputed"));
  }, a = ({
    detail: l
  }) => {
    const { cell: d } = l, { Name: m, Value: u } = d.rowData;
    m && u === void 0 && setTimeout(() => {
      d.style.gridColumn = "1 / -1";
    });
  };
  return c.html`
    <bim-table @cellcreated=${a} ${c.ref(s)}>
      ${n ? c.html`
            <bim-label slot="missing-data" style="--bim-icon--c: gold" icon="ic:round-warning">
              Select some elements to display its properties
            </bim-label>
            ` : null}
      <bim-label slot="error-loading" style="--bim-icon--c: #e72e2e" icon="bxs:error-alt">
        Something went wrong with the properties
      </bim-label>
    </bim-table>
  `;
}, ce = /* @__PURE__ */ new Map(), de = {
  METRE: "m",
  SQUARE_METRE: "m²",
  CUBIC_METRE: "m³"
}, me = async (r, t) => {
  const i = r.get(x.FragmentsManager).list.get(t);
  if (!i)
    throw new Error(`ItemsDataUI: model ${t} not found.`);
  let n = ce.get(i.modelId);
  if (!n) {
    const [o] = Object.values(await i.getItemsOfCategories([/UNITASSIGNMENT/])).flat(), [s] = await i.getItemsData([o], {
      // Units is the relation from the IfcUnitAssignment that holds the list of units in the project
      relations: { Units: { relations: !1, attributes: !0 } }
    });
    if (!Array.isArray(s.Units))
      return [];
    n = s.Units;
  }
  return n;
}, ue = (r, t) => {
  const { components: e } = r;
  t.columns = [{ name: "Name", width: "12rem" }], t.visibleColumns = ["Name", "Value"], t.headersHidden = !0, t.dataTransform = {
    Value: (i, n) => {
      const { dataType: o, modelId: s } = n;
      if (!o)
        return i;
      const a = async (l) => {
        if (!(l && s))
          return;
        const d = await me(e, s), m = o.replace("IFC", "").replace("MEASURE", "UNIT"), u = d.find((b) => b.UnitType && "value" in b.UnitType ? b.UnitType.value === m : !1);
        if (!u || !(u.Name && "value" in u.Name))
          return i;
        const p = `${i.toFixed(2)} ${de[u.Name.value] ?? u.Name.value}`;
        l.textContent = p;
      };
      return c.html`<bim-label ${c.ref(a)}></bim-label>`;
    }
  };
}, pe = (r) => {
  const t = c.Component.create(vt, r), [e] = t;
  return ue(r, e), t;
}, he = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  itemsData: pe,
  itemsDataTemplate: vt
}, Symbol.toStringTag, { value: "Module" })), wt = (r) => {
  const { components: t } = r, e = r.missingDataMessage ?? "No models has been loaded yet", i = t.get(x.FragmentsManager), n = ({
    detail: s
  }) => {
    const { cell: a } = s;
    a.style.padding = "0.25rem 0";
  }, o = async (s) => {
    if (!s)
      return;
    const a = s, l = [];
    if (i.initialized)
      for (const [, d] of i.list) {
        if (!(d && !d.isDeltaModel))
          continue;
        const m = await d.getMetadata(), u = {
          data: {
            Name: d.modelId,
            modelId: d.modelId,
            metadata: JSON.stringify(m)
          }
        };
        l.push(u);
      }
    a.data = l;
  };
  return c.html`
    <bim-table ${c.ref(o)} @cellcreated=${n}>
      <bim-label slot="missing-data" style="--bim-icon--c: gold" icon="ic:round-warning">
        ${e}
      </bim-label>
    </bim-table>
  `;
}, be = (r, t) => {
  const { components: e, actions: i, metaDataTags: n } = r, o = e.get(x.FragmentsManager), s = (i == null ? void 0 : i.dispose) ?? !0, a = (i == null ? void 0 : i.download) ?? !0, l = (i == null ? void 0 : i.visibility) ?? !0, d = n ?? [];
  t.hiddenColumns = ["modelId", "metadata"], t.headersHidden = !0, t.noIndentation = !0, t.dataTransform = {
    Name: (m, u) => {
      if (!o.initialized)
        return m;
      const { modelId: p, metadata: b } = u;
      if (!p)
        return m;
      const h = o.list.get(p);
      if (!h)
        return p;
      const f = [];
      if (b) {
        const y = JSON.parse(b);
        for (const v of d) {
          const $ = y[v];
          if (!(typeof $ == "string" || typeof $ == "boolean" || typeof $ == "number"))
            continue;
          const S = c.html`
            <bim-label style="background-color: var(--bim-ui_main-base); padding: 0 0.25rem; color: var(--bim-ui_main-contrast); border-radius: 0.25rem;">${$}</bim-label>
            `;
          f.push(S);
        }
      }
      let w;
      if (s) {
        const y = () => o.core.disposeModel(h.modelId);
        w = c.html`<bim-button @click=${y} icon="mdi:delete"></bim-button>`;
      }
      let _;
      if (l) {
        const y = async ({ target: v }) => {
          v.loading = !0, await h.setVisible(
            void 0,
            v.hasAttribute("data-model-hidden")
          ), await o.core.update(!0), v.toggleAttribute("data-model-hidden"), v.icon = v.hasAttribute("data-model-hidden") ? "mdi:eye-off" : "mdi:eye", v.loading = !1;
        };
        _ = c.html`<bim-button @click=${y} icon="mdi:eye"></bim-button>`;
      }
      let g;
      if (a) {
        const y = async () => {
          const v = await h.getBuffer(!1), $ = new File([v], `${h.modelId}.frag`), S = document.createElement("a");
          S.href = URL.createObjectURL($), S.download = $.name, S.click(), URL.revokeObjectURL(S.href);
        };
        g = c.html`<bim-button @click=${y} icon="flowbite:download-solid"></bim-button>`;
      }
      return c.html`
       <div style="display: flex; flex: 1; gap: var(--bim-ui_size-4xs); justify-content: space-between; overflow: auto;">
        <div style="display: flex; flex-wrap: wrap; align-items: center; gap: 0 var(--bim-ui_size-4xs); flex-grow: 1; overflow: auto;">
          <div style="min-height: 1.75rem; overflow: auto; display: flex;">
            <bim-label style="white-space: normal;">${m}</bim-label>
          </div>
          <div style="display: flex; flex-wrap: wrap; gap: var(--bim-ui_size-4xs); overflow: auto;">
            ${f}
          </div>
        </div>
        <div style="display: flex; align-self: flex-start; flex-shrink: 0;">
          ${g}
          ${_}
          ${w}
        </div>
       </div>
      `;
    }
  };
}, fe = (r, t = !0) => {
  const e = c.Component.create(wt, r), [i, n] = e;
  if (be(r, i), t) {
    const { components: o } = r, s = o.get(x.FragmentsManager), a = () => setTimeout(() => n());
    s.list.onItemSet.add(a), s.list.onItemDeleted.add(a);
  }
  return e;
}, ge = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  modelsList: fe,
  modelsListTemplate: wt
}, Symbol.toStringTag, { value: "Module" })), xt = (r) => {
  var l;
  const { components: t } = r, e = r.missingDataMessage ?? "No viewpoints to display", i = t.get(x.Viewpoints), n = ((l = r.topic) == null ? void 0 : l.viewpoints) ?? i.list.keys(), o = [];
  for (const d of n) {
    const m = i.list.get(d);
    m && o.push(m);
  }
  const s = (d) => {
    if (!d)
      return;
    const m = d;
    m.data = o.map((u, p) => ({
      data: {
        Guid: u.guid,
        Title: u.title ?? `Viewpoint ${r.topic ? p + 1 : ""}`,
        Actions: ""
      }
    }));
  }, a = ({
    detail: d
  }) => {
    const { cell: m } = d;
    m.style.padding = "0.25rem";
  };
  return c.html`
    <bim-table ${c.ref(s)} @cellcreated=${a}>
      <bim-label slot="missing-data" icon="ph:warning-fill" style="--bim-icon--c: gold;">${e}</bim-label>
    </bim-table>
  `;
}, ye = (r, t) => {
  const { components: e, topic: i } = r;
  t.noIndentation = !0, t.headersHidden = !0, t.hiddenColumns = ["Guid"], t.columns = ["Title", { name: "Actions", width: "auto" }];
  const n = {
    selectComponents: !0,
    colorizeComponent: !0,
    resetColors: !0,
    updateCamera: !0,
    delete: !0,
    unlink: !!i,
    ...r.actions
  }, o = e.get(x.Viewpoints);
  t.dataTransform = {
    Actions: (s, a) => {
      const { Guid: l } = a;
      if (!(l && typeof l == "string"))
        return s;
      const d = o.list.get(l);
      if (!d)
        return s;
      const m = async ({ target: g }) => {
        g.loading = !0, await d.go(), g.loading = !1;
      };
      let u;
      if (n.selectComponents) {
        const g = async ({ target: y }) => {
          const v = e.get(x.FragmentsManager), $ = e.get(q.Highlighter);
          if (!$.isSetup)
            return;
          y.loading = !0;
          const S = await v.guidsToModelIdMap([
            ...d.selectionComponents
          ]);
          await $.highlightByID("select", S), y.loading = !1;
        };
        u = c.html`
          <bim-button label="Select Components" @click=${g}></bim-button>
        `;
      }
      let p;
      if (n.colorizeComponent) {
        const g = async ({ target: y }) => {
          y.loading = !0, await d.setColorizationState(!0), y.loading = !1;
        };
        p = c.html`
          <bim-button label="Colorize Components" @click=${g}></bim-button>
        `;
      }
      let b;
      if (n.resetColors) {
        const g = async ({ target: y }) => {
          y.loading = !0, await d.setColorizationState(!1), y.loading = !1;
        };
        b = c.html`
          <bim-button label="Reset Colors" @click=${g}></bim-button>
        `;
      }
      let h;
      if (n.updateCamera) {
        const g = () => d.updateCamera();
        h = c.html`
          <bim-button label="Update Camera" @click=${g}></bim-button>
        `;
      }
      let f;
      if (n.unlink) {
        const g = () => i == null ? void 0 : i.viewpoints.delete(d.guid);
        f = c.html`
          <bim-button label="Unlink" @click=${g}></bim-button>
        `;
      }
      let w;
      if (n.delete) {
        const g = () => {
          o.list.delete(d.guid), c.ContextMenu.removeMenus();
        };
        w = c.html`
          <bim-button label="Delete" @click=${g}></bim-button>
        `;
      }
      let _;
      return Object.values(n).includes(!0) && (_ = c.html`
          <bim-button icon="prime:ellipsis-v">
            <bim-context-menu>
              ${u}
              ${p}
              ${b}
              ${h}
              ${f}
              ${w}
            </bim-context-menu>
          </bim-button>
        `), c.html`
        <bim-button icon="ph:eye-fill" @click=${m}></bim-button>
        ${_}
      `;
    }
  };
}, Ct = (r, t = !0) => {
  const e = c.Component.create(xt, r), [i, n] = e;
  if (ye(r, i), t) {
    const { components: o, topic: s } = r, a = o.get(x.Viewpoints);
    a.list.onItemUpdated.add(() => n()), a.list.onItemDeleted.add(() => n()), a.list.onCleared.add(() => n()), s ? (s.viewpoints.onItemAdded.add(() => n()), s.viewpoints.onItemDeleted.add(() => n()), s.viewpoints.onCleared.add(() => n())) : a.list.onItemSet.add(() => n());
  }
  return e;
}, _e = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  viewpointsList: Ct,
  viewpointsListTemplate: xt
}, Symbol.toStringTag, { value: "Module" })), $t = (r) => {
  const { components: t } = r, e = r.missingDataMessage ?? "No topics to display", i = t.get(x.BCFTopics), n = r.topics ?? i.list.values(), o = (s) => {
    if (!s)
      return;
    const a = s;
    a.data = [...n].map((l) => {
      var d;
      return {
        data: {
          Guid: l.guid,
          Title: l.title,
          Status: l.status,
          Description: l.description ?? "",
          Author: l.creationAuthor,
          Assignee: l.assignedTo ?? "",
          Date: l.creationDate.toDateString(),
          DueDate: ((d = l.dueDate) == null ? void 0 : d.toDateString()) ?? "",
          Type: l.type,
          Priority: l.priority ?? "",
          Actions: ""
        }
      };
    });
  };
  return c.html`
    <bim-table no-indentation ${c.ref(o)}>
      <bim-label slot="missing-data" icon="ph:warning-fill" style="--bim-icon--c: gold;">${e}</bim-label>
    </bim-table>
  `;
}, ve = (r, t) => {
  const { dataStyles: e } = r;
  t.hiddenColumns.length === 0 && (t.hiddenColumns = ["Guid", "Actions"]), t.columns = ["Title"], t.dataTransform = {
    Priority: (i) => {
      if (typeof i != "string")
        return i;
      const o = ((e == null ? void 0 : e.priorities) ?? E.priorities)[i];
      return c.html`
            <bim-label
              .icon=${o == null ? void 0 : o.icon}
              style=${c.styleMap({ ...H, ...o == null ? void 0 : o.style })}
            >${i}
            </bim-label>
          `;
    },
    Status: (i) => {
      if (typeof i != "string")
        return i;
      const o = ((e == null ? void 0 : e.statuses) ?? E.statuses)[i];
      return c.html`
            <bim-label
              .icon=${o == null ? void 0 : o.icon}
              style=${c.styleMap({ ...H, ...o == null ? void 0 : o.style })}
            >${i}
            </bim-label>
          `;
    },
    Type: (i) => {
      if (typeof i != "string")
        return i;
      const o = ((e == null ? void 0 : e.types) ?? E.types)[i];
      return c.html`
            <bim-label
              .icon=${o == null ? void 0 : o.icon}
              style=${c.styleMap({ ...H, ...o == null ? void 0 : o.style })}
            >${i}
            </bim-label>
          `;
    },
    Author: (i) => typeof i != "string" ? i : G(i, (e == null ? void 0 : e.users) ?? E.users),
    Assignee: (i) => typeof i != "string" ? i : G(i, (e == null ? void 0 : e.users) ?? E.users)
  };
}, St = (r, t = !0) => {
  const e = c.Component.create($t, r), [i, n] = e;
  if (ve(r, i), t) {
    const { components: o, topics: s } = r, a = o.get(x.BCFTopics), l = () => n();
    if (a.list.onItemUpdated.add(l), a.list.onItemDeleted.add(l), s)
      for (const d of s)
        d.relatedTopics.onItemAdded.add(l), d.relatedTopics.onItemDeleted.add(l), d.relatedTopics.onCleared.add(l);
    else
      a.list.onItemSet.add(l);
  }
  return e;
}, we = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  topicsList: St,
  topicsListTemplate: $t
}, Symbol.toStringTag, { value: "Module" })), Tt = (r) => {
  const { topic: t, styles: e, viewpoint: i } = r, n = r.missingDataMessage ?? "The topic has no comments", o = (s) => {
    if (!s)
      return;
    const a = s;
    let l = t.comments.values();
    i && (l = [...t.comments.values()].filter(
      (d) => d.viewpoint === i.guid
    )), a.data = [...l].map((d) => ({
      data: {
        guid: d.guid,
        Comment: d.comment,
        author: (() => {
          const m = e;
          if (!m)
            return d.author;
          const u = m[d.author];
          return (u == null ? void 0 : u.name) ?? d.author;
        })()
      }
    }));
  };
  return c.html`
    <bim-table no-indentation ${c.ref(o)}>
      <bim-label slot="missing-data" icon="ph:warning-fill" style="--bim-icon--c: gold;">${n}</bim-label>
    </bim-table>
  `;
}, xe = (r, t) => {
  const { topic: e, styles: i } = r, n = { delete: !0, ...r.actions };
  t.headersHidden = !0, t.hiddenColumns = ["guid", "author"], t.dataTransform = {
    Comment: (o, s) => {
      const { guid: a } = s;
      if (typeof a != "string")
        return o;
      const l = e.comments.get(a);
      if (!l)
        return o;
      const d = () => {
        e.comments.delete(a);
      };
      let m;
      if (n.delete) {
        const u = `btn-${c.Manager.newRandomId()}`;
        m = c.html`
          <div>
            <style>
              #${u} {
                background-color: transparent;
                --bim-label--c: var(--bim-ui_bg-contrast-60)
              }
  
              #${u}:hover {
                --bim-label--c: #FF5252;
              }
            </style>
            <bim-button @click=${d} id=${u} icon="majesticons:delete-bin"></bim-button>
          </div>
        `;
      }
      return c.html`
        <div style="display: flex; flex-direction: column; gap: 0.25rem; flex: 1">
          <div style="display: flex; justify-content: space-between;">
            <div style="display: flex; gap: 0.375rem;">
              ${G(l.author, i ?? E.users)}
              <bim-label style="color: var(--bim-ui_bg-contrast-40)">@ ${l.date.toDateString()}</bim-label>
            </div>
            ${m}
          </div>
          <bim-label style="margin-left: 1.7rem; white-space: normal">${l.comment}</bim-label>
        </div>
      `;
    }
  };
}, Mt = (r, t = !0) => {
  const e = c.Component.create(Tt, r), [i, n] = e;
  if (xe(r, i), t) {
    const { topic: o } = r, s = () => n();
    o.comments.onItemSet.add(s), o.comments.onItemUpdated.add(s), o.comments.onItemDeleted.add(s), o.comments.onCleared.add(s);
  }
  return e;
}, Ce = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  commentsList: Mt,
  commentsListTemplate: Tt
}, Symbol.toStringTag, { value: "Module" })), mi = {
  // ...worldsConfiguration,
  ...re,
  ...he,
  ...ge,
  ..._e,
  ...we,
  ...Ce
}, It = (r, t) => {
  const { showInput: e, topic: i, styles: n } = r, o = {
    add: !0,
    delete: !0,
    ...r.actions
  }, s = `input-${c.Manager.newRandomId()}`, a = `btn-${c.Manager.newRandomId()}`, l = `btn-${c.Manager.newRandomId()}`, d = () => document.getElementById(a), m = () => document.getElementById(s), u = () => {
    const v = m();
    return v ? v.value.trim().length > 0 : !1;
  }, p = () => {
    t({ showInput: !0 });
  }, b = () => {
    const v = m(), $ = u();
    v && $ && (i.createComment(v.value), t({ showInput: !1 }));
  }, h = () => {
    t({ showInput: !1 });
  }, f = () => {
    const v = d();
    if (!v)
      return;
    if (!m()) {
      v.disabled = !0;
      return;
    }
    v.disabled = !u();
  }, w = c.html`
    ${o.add ? c.html`<bim-button @click=${p} label="Add Comment" icon="majesticons:comment-line"></bim-button>` : null}
  `, _ = (v) => {
    v.code === "Enter" && v.ctrlKey && b();
  }, g = c.html`
    <bim-text-input id=${s} @input=${f} @keypress=${_} type="area"></bim-text-input>

    <div style="justify-content: right; display: flex; gap: 0.375rem">
      <style>
        #${a} {
          background-color: #329936;
        }  

        #${l} {
          background-color: transparent;
        }

        #${l}:hover {
          --bim-label--c: #FF5252;
        }
      </style>

      <bim-button style="flex: 0" id=${l} @click=${h} label="Cancel"></bim-button>
      <bim-button style="flex: 0" id=${a} @click=${b} label="Accept" icon="material-symbols:check" disabled></bim-button>
    </div>
  `, [y] = Mt({
    topic: i,
    actions: o,
    styles: n ?? E.users
  });
  return c.html`
    <div style="display: flex; flex-direction: column; gap: 0.5rem">
      ${y}
      ${e ? g : w}
    </div>
  `;
}, $e = (r) => c.Component.create(It, r), Se = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  topicComments: $e,
  topicCommentsSectionTemplate: It
}, Symbol.toStringTag, { value: "Module" })), Dt = (r, t) => {
  const { components: e, editing: i, topic: n, styles: o } = r, s = {
    update: !0,
    ...r.actions
  }, a = (o == null ? void 0 : o.priorities) ?? E.priorities, l = (o == null ? void 0 : o.statuses) ?? E.statuses, d = (o == null ? void 0 : o.types) ?? E.types;
  let m;
  n != null && n.priority && (m = a[n.priority]);
  let u;
  n != null && n.type && (u = d[n.type]);
  let p;
  n != null && n.type && (p = l[n.status]);
  let b, h;
  return i ? b = gt({
    components: e,
    topic: n,
    styles: o,
    onSubmit: () => {
      t({ editing: !1 });
    },
    onCancel: () => {
      t({ editing: !1 });
    }
  }) : h = c.html`
      <div>
        <bim-label>Title</bim-label>
        <bim-label style="--bim-label--c: var(--bim-ui_bg-contrast-100)">${n.title}</bim-label>
      </div>

      ${n.description ? c.html`
            <div>
              <bim-label>Description</bim-label>
              <bim-label style="--bim-label--c: var(--bim-ui_bg-contrast-100); white-space: normal">${n.description}</bim-label>
            </div>
          ` : null}

      <div style="display: flex; gap: 0.375rem">
        <bim-label>Status</bim-label>
        <bim-label .icon=${p == null ? void 0 : p.icon} style=${c.styleMap({ ...H, ...p == null ? void 0 : p.style })}
        >${n.status}
        </bim-label>
      </div>

      <div style="display: flex; gap: 0.375rem">
        <bim-label>Type</bim-label>
        <bim-label .icon=${u == null ? void 0 : u.icon} style=${c.styleMap({ ...H, ...u == null ? void 0 : u.style })}
        >${n.type}
        </bim-label>
      </div>

      ${n.priority ? c.html`
            <div style="display: flex; gap: 0.375rem">
              <bim-label>Priority</bim-label>
              <bim-label .icon=${m == null ? void 0 : m.icon} style=${c.styleMap({ ...H, ...m == null ? void 0 : m.style })}
              >${n.priority}
              </bim-label>
            </div>` : null}

      <div style="display: flex; gap: 0.375rem">
        <bim-label>Author</bim-label>
        ${G(n.creationAuthor, (o == null ? void 0 : o.users) ?? E.users)}
      </div>

      ${n.assignedTo ? c.html`
          <div style="display: flex; gap: 0.375rem">
            <bim-label>Assignee</bim-label>
            ${G(n.assignedTo, (o == null ? void 0 : o.users) ?? E.users)}
          </div>` : null}

      ${n.dueDate ? c.html`
          <div style="display: flex; gap: 0.375rem">
            <bim-label>Due Date</bim-label>
            <bim-label style="--bim-label--c: var(--bim-ui_bg-contrast-100)">${n.dueDate.toDateString()}</bim-label>
          </div>` : null}

      ${n.modifiedAuthor ? c.html`
          <div style="display: flex; gap: 0.375rem">
            <bim-label>Modified By</bim-label>
            ${G(n.modifiedAuthor, (o == null ? void 0 : o.users) ?? E.users)}
          </div>` : null}

      ${n.modifiedDate ? c.html`
            <div style="display: flex; gap: 0.375rem">
              <bim-label>Modified Date</bim-label>
              <bim-label style="--bim-label--c: var(--bim-ui_bg-contrast-100)">${n.modifiedDate.toDateString()}</bim-label>
            </div>` : null}

      ${n.labels.size !== 0 ? c.html`
          <div style="display: flex; gap: 0.375rem">
            <bim-label>Labels</bim-label>
            <bim-label style="white-space: normal; --bim-label--c: var(--bim-ui_bg-contrast-100)">${[...n.labels].join(", ")}</bim-label>
          </div>` : null}

      ${s.update ? c.html`
              <bim-button @click=${() => t({ editing: !0 })} label="Update Information" icon="tabler:refresh"></bim-button> 
            ` : null}
    `, c.html`
    <div style="display: flex; flex-direction: column; gap: 0.75rem;">
      ${i ? b : h}
    </div>
  `;
}, Te = (r) => c.Component.create(Dt, r), Me = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  topicInformation: Te,
  topicInformationSectionTemplate: Dt
}, Symbol.toStringTag, { value: "Module" })), kt = (r, t) => {
  const { components: e, topic: i, linking: n } = r, o = e.get(x.BCFTopics), s = {
    link: !0,
    ...r.actions
  }, [a, l] = St({
    components: e,
    topics: [...i.relatedTopics].map((p) => o.list.get(p)).map((p) => p)
  });
  a.headersHidden = !0, a.hiddenColumns = [
    "Guid",
    "Status",
    // "Title",
    "Description",
    "Author",
    "Assignee",
    "Date",
    "DueDate",
    "Type",
    "Priority"
  ];
  const d = () => {
    const p = (b) => {
      const h = b.target;
      h instanceof c.TextInput && (a.queryString = h.value);
    };
    return c.html`
      <bim-text-input placeholder="Search..." debounce="100" @input=${p}></bim-text-input> 
    `;
  };
  let m, u;
  if (n) {
    a.selectableRows = !0, l({
      topics: void 0
    });
    const p = a.data.filter((_) => {
      const { Guid: g } = _.data;
      return typeof g != "string" ? !1 : i.relatedTopics.has(g);
    }).map((_) => _.data);
    a.selection.add(...p);
    const b = () => {
      const _ = [...a.selection].map(({ Guid: g }) => typeof g != "string" ? null : o.list.has(g) ? g : null).map((g) => g);
      i.relatedTopics.clear(), i.relatedTopics.add(..._), t({ linking: !1 });
    }, h = () => {
      t({ linking: !1 });
    }, f = `btn-${c.Manager.newRandomId()}`, w = `btn-${c.Manager.newRandomId()}`;
    m = c.html`
      <div style="display: flex; gap: 0.25rem">
        <style>
          #${f}:hover {
            background-color: #329936;
          }  

          #${w} {
            background-color: transparent;
          }

          #${w}:hover {
            --bim-label--c: #FF5252;
          }
        </style>
        ${d()}
        <div style="display: flex; justify-content: right; gap: 0.25rem">
          <bim-button id=${w} @click=${h} style="flex: 0" label="Cancel" icon="material-symbols:close"></bim-button>
          <bim-button id=${f} @click=${b} style="flex: 0" label="Accept" icon="material-symbols:check"></bim-button>
        </div>
      </div> 
    `;
  } else {
    a.selectableRows = !1;
    const p = () => {
      t({ linking: !0 });
    };
    u = c.html`
      <div style="display: flex; justify-content: right; gap: 0.25rem">
        ${d()}
        ${s.link ? c.html`<bim-button style="flex: 0" @click=${p} icon="tabler:link"></bim-button>` : null}
      </div> 
    `;
  }
  return c.html`
    <div style="display: flex; flex-direction: column; gap: 0.5rem;">
      ${u}
      ${m}
      ${a}
    </div> 
  `;
}, Ie = (r) => c.Component.create(kt, r), De = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  topicRelations: Ie,
  topicRelationsSectionTemplate: kt
}, Symbol.toStringTag, { value: "Module" })), Et = (r, t) => {
  const { components: e, topic: i, world: n, linking: o } = r, s = {
    add: !0,
    link: !0,
    selectComponents: !0,
    colorizeComponent: !0,
    resetColors: !0,
    updateCamera: !0,
    delete: !0,
    unlink: !0,
    ...r.actions
  }, a = e.get(x.Viewpoints), [l, d] = Ct({
    components: e,
    topic: i,
    actions: s
  }), m = () => {
    const b = (h) => {
      const f = h.target;
      f instanceof c.TextInput && (l.queryString = f.value);
    };
    return c.html`
      <bim-text-input placeholder="Search..." debounce="100" @input=${b}></bim-text-input> 
    `;
  };
  let u, p;
  if (o) {
    l.selectableRows = !0, d({
      topic: void 0,
      actions: {
        delete: !1,
        updateCamera: !1,
        colorizeComponent: !1,
        resetColors: !1
      }
    });
    const b = l.data.filter((g) => {
      const { Guid: y } = g.data;
      return typeof y != "string" ? !1 : i.viewpoints.has(y);
    }).map((g) => g.data);
    l.selection.add(...b);
    const h = () => {
      const g = [...l.selection].map(({ Guid: y }) => typeof y != "string" ? null : a.list.has(y) ? y : null).map((y) => y);
      i.viewpoints.clear(), i.viewpoints.add(...g), t({ linking: !1 });
    }, f = () => {
      t({ linking: !1 });
    }, w = `btn-${c.Manager.newRandomId()}`, _ = `btn-${c.Manager.newRandomId()}`;
    u = c.html`
      <div style="display: flex; gap: 0.25rem">
        <style>
          #${w}:hover {
            background-color: #329936;
          }  

          #${_} {
            background-color: transparent;
          }

          #${_}:hover {
            --bim-label--c: #FF5252;
          }
        </style>
        ${m()}
        <div style="display: flex; justify-content: right; gap: 0.25rem">
          <bim-button id=${_} @click=${f} style="flex: 0" label="Cancel" icon="material-symbols:close"></bim-button>
          <bim-button id=${w} @click=${h} style="flex: 0" label="Accept" icon="material-symbols:check"></bim-button>
        </div>
      </div> 
    `;
  } else {
    l.selectableRows = !1, d({
      topic: i,
      actions: s
    });
    const b = () => {
      if (!(i && s.add && !o))
        return;
      const _ = a.create();
      n && (_.world = n), i.viewpoints.add(_.guid);
    }, h = () => {
      t({ linking: !0 });
    }, f = c.html`<bim-button style="flex: 0" @click=${b} .disabled=${!n} icon="mi:add"></bim-button>`, w = c.html`<bim-button style="flex: 0" @click=${h} icon="tabler:link"></bim-button>`;
    p = c.html`
      <div style="display: flex; justify-content: right; gap: 0.25rem">
        ${m()}
        <div style="display: flex; justify-content: right; gap: 0.25rem">
          ${s.add ? f : null}
          ${s.link ? w : null}
        </div>
      </div> 
    `;
  }
  return c.html`
    <div style="display: flex; flex-direction: column; gap: 0.5rem;">
      ${p}
      ${u}
      ${l}
    </div> 
  `;
}, ke = (r) => c.Component.create(Et, r), Ee = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  topicViewpoints: ke,
  topicViewpointsSectionTemplate: Et
}, Symbol.toStringTag, { value: "Module" })), ui = {
  ...Se,
  ...Me,
  ...De,
  ...Ee
  // ...specificationInformation,
}, Re = (r) => c.html`
    <bim-panel-section fixed label="New Topic" name="topic">
      ${gt(r)}
    </bim-panel-section>
  `, Pe = (r) => c.Component.create(
  Re,
  r
), Oe = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  topic: Pe
}, Symbol.toStringTag, { value: "Module" })), pi = {
  ...Oe
};
async function Le(r, t) {
  const e = t.get(x.FragmentsManager), i = new x.DataMap();
  for (const [s, a] of e.list)
    for (const [l, d] of Object.entries(r)) {
      const m = await a.getItemsData(Array.from(d));
      for (const u of m) {
        if (!u || !("value" in u._category))
          continue;
        const p = u._category.value, b = i.get(p) ?? 0;
        i.set(p, b + 1);
      }
    }
  const n = [...i.keys()], o = [...i.values()].map((s) => ({ value: s }));
  return {
    labels: n,
    datasets: {
      Categories: o
    }
  };
}
const ze = (r) => {
  const { type: t, modelIdMap: e, components: i, addLabels: n = !0 } = r, o = i.get(x.FragmentsManager), s = i.get(q.Highlighter), a = i.get(x.Hider), l = r.missingDataMessage ?? "No data in this chart.", d = async (m) => {
    if (!m)
      return;
    const u = m;
    if (u.loadFunction = async () => Le(e, i), await u.loadData(!0), n) {
      const p = u.querySelector(
        "[slot='labels']"
      );
      p && (p.charts = [u], p.addEventListener("label-click", async (b) => {
        const h = b, { label: f, visibility: w } = h.detail, _ = [];
        for (const [g, y] of o.list) {
          const v = await y.getItemsOfCategories([new RegExp(f)]), $ = Object.values(v).flat(), S = { [g]: new Set($) };
          _.push(a.set(w, S));
        }
        _.push(o.core.update(!0)), await Promise.all(_);
      }));
    }
    u.addEventListener("sectionclick", async (p) => {
      const b = p, { label: h } = b.detail;
      for (const [f, w] of o.list) {
        const _ = await w.getItemsOfCategories([
          new RegExp(h)
        ]), g = Object.values(_).flat(), y = { [f]: new Set(g) };
        s.highlightByID("select", y, !0, !0);
      }
    });
  };
  return c.html`
    <bim-chart ${c.ref(d)} type=${t ?? "bar"}>
      <bim-label slot="missing-data" icon="ph:warning-fill" style="--bim-icon--c: gold;">${l}</bim-label>
      ${n ? c.html`<bim-chart-legend slot="labels"></bim-chart-legend>` : ""}
    </bim-chart>
  `;
}, Fe = (r) => c.Component.create(
  ze,
  r
), Ae = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  categoriesChart: Fe
}, Symbol.toStringTag, { value: "Module" })), Rt = "Passed", Pt = "Failed";
async function Be(r) {
  const t = [], e = [], i = {};
  for (const s of r)
    for (const [a, l] of s.entries()) {
      for (const d of l.values()) {
        const { guid: m, pass: u } = d;
        m && (u ? t.push(m) : e.push(m));
      }
      i[a] = { Passed: t, Failed: e };
    }
  const n = {
    value: t.length,
    data: { guids: t, guidsMap: i }
  }, o = {
    value: e.length,
    data: { guids: e, guidsMap: i }
  };
  return {
    labels: [Rt, Pt],
    datasets: {
      "IDS Compliance": [n, o]
    }
  };
}
const je = (r) => {
  const { type: t, idsResult: e, components: i, addLabels: n = !0 } = r, o = i.get(x.FragmentsManager), s = i.get(q.Highlighter), a = i.get(x.Hider);
  s.styles.set(Rt, {
    color: new I.Color("green"),
    renderedFaces: bt.RenderedFaces.ONE,
    opacity: 1,
    transparent: !1
  }), s.styles.set(Pt, {
    color: new I.Color("red"),
    renderedFaces: bt.RenderedFaces.ONE,
    opacity: 1,
    transparent: !1
  });
  const l = r.missingDataMessage ?? "No data in this chart.", d = async (m) => {
    if (!m)
      return;
    const u = m;
    if (u.loadFunction = async () => {
      const p = Array.isArray(e) ? e : [e];
      return Be(p);
    }, u.colors = ["green", "red"], await u.loadData(!0), n) {
      const p = u.querySelector(
        "[slot='labels']"
      );
      p && (p.charts = [u]), p.addEventListener(
        "label-click",
        async (b) => {
          const { label: h, data: f, visibility: w } = b.detail, _ = [];
          for (const g of f) {
            const { guidsMap: y } = g;
            for (const [v, $] of Object.entries(y)) {
              const S = o.list.get(v);
              if (!S)
                continue;
              const O = $[h];
              if (!O || O.length === 0)
                continue;
              const F = (async () => {
                const A = (await S.getLocalIdsByGuids(O)).filter(
                  (L) => L !== null
                );
                if (A.length === 0)
                  return;
                const B = { [v]: new Set(A) };
                await a.set(w, B);
              })();
              _.push(F);
            }
          }
          await Promise.all(_);
        }
      );
    }
    u.addEventListener(
      "sectionclick",
      async (p) => {
        const { values: b, label: h } = p.detail;
        await s.clear();
        for (const f of b) {
          const { data: w } = f;
          if (!w)
            continue;
          const { guids: _ } = w;
          for (const [g, y] of o.list) {
            const v = (await y.getLocalIdsByGuids(_)).filter(
              (S) => S !== null
            ), $ = { [g]: new Set(v) };
            await s.highlightByID(h, $, !0, !1);
          }
        }
      }
    );
  };
  return c.html`
    <bim-chart ${c.ref(d)} type=${t ?? "bar"} colorfulBars=${t === "bar"}>
      <bim-label slot="missing-data" icon="ph:warning-fill" style="--bim-icon--c: gold;">${l}</bim-label>
      ${n ? c.html`<bim-chart-legend slot="labels"></bim-chart-legend>` : ""}
    </bim-chart>
  `;
}, Ne = (r) => c.Component.create(
  je,
  r
), Ve = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  idsChart: Ne
}, Symbol.toStringTag, { value: "Module" })), Ue = (r) => {
  const { charts: t } = r, e = r.missingDataMessage ?? "No charts attached", i = async (n) => {
    if (!n)
      return;
    const o = n;
    o.charts = t;
  };
  return c.html`
    <bim-chart-legend ${c.ref(i)}>
      <bim-label slot="no-chart" icon="ph:warning-fill" style="--bim-icon--c: gold;">${e}</bim-label>
    </bim-chart-legend>
  `;
}, Xe = (r) => c.Component.create(
  Ue,
  r
), Ye = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  labels: Xe
}, Symbol.toStringTag, { value: "Module" }));
async function Ge(r, t) {
  const e = {};
  for (const o of r) {
    let s = o[t];
    s == null ? s = "Not defined" : s instanceof Date ? s = s.toLocaleDateString() : s = String(s), e[s] || (e[s] = []), e[s].push(o);
  }
  const i = Object.keys(e), n = [];
  for (const o of i) {
    const s = e[o];
    n.push({
      value: s.length,
      data: { topics: s.map(({ guid: a }) => a) }
    });
  }
  return {
    labels: i,
    datasets: {
      Topics: n
    }
  };
}
const He = (r) => {
  const { components: t, type: e, addLabels: i = !0, grouper: n = "stage" } = r, o = r.filterFunction ?? (() => !0), a = [...t.get(x.BCFTopics).list.values()].filter(o), l = r.missingDataMessage ?? "No data in this chart.", d = async (m) => {
    if (!m)
      return;
    const u = m;
    if (u.loadFunction = async () => Ge(a, n), await u.loadData(!0), i) {
      const p = u.querySelector(
        "[slot='labels']"
      );
      p && (p.charts = [u]);
    }
  };
  return c.html`
    <bim-chart ${c.ref(d)} type=${e ?? "bar"}>
      <bim-label slot="missing-data" icon="ph:warning-fill" style="--bim-icon--c: gold;">${l}</bim-label>
      ${i ? c.html`<bim-chart-legend slot="labels"></bim-chart-legend>` : ""}
    </bim-chart>
  `;
}, qe = (r) => c.Component.create(
  He,
  r
), We = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  topicsChart: qe
}, Symbol.toStringTag, { value: "Module" }));
async function Je(r, t, e, i) {
  const o = i.get(x.FragmentsManager).list.get(e);
  if (!o)
    return { labels: [], datasets: {} };
  const s = await o.getAttributesUniqueValues([
    {
      categories: [t],
      get: r
    }
  ]);
  if (!s)
    return { labels: [], datasets: {} };
  const a = [], l = [];
  for (const [d, m] of Object.entries(s))
    for (const { value: u, localIds: p } of Object.values(m)) {
      a.push(u);
      const b = p.length;
      l.push({
        value: b,
        data: { modelIdMap: { [e]: new Set(p) } }
      });
    }
  return {
    labels: a,
    datasets: {
      [r.toString()]: l
    }
  };
}
const Ke = (r) => {
  const {
    type: t,
    attribute: e,
    category: i,
    modelId: n,
    components: o,
    addLabels: s = !0
  } = r, a = { zoom: !0, ...r.highlight }, l = o.get(x.FragmentsManager), d = o.get(q.Highlighter), m = o.get(x.Hider), u = r.missingDataMessage ?? "No data in this chart.", p = async (b) => {
    if (!b)
      return;
    const h = b;
    if (h.loadFunction = async () => Je(e, i, n, o), await h.loadData(!0), s) {
      const f = h.querySelector(
        "[slot='labels']"
      );
      f && (f.charts = [h]), f.addEventListener(
        "label-click",
        async (w) => {
          const { data: _, visibility: g } = w.detail;
          for (const y of _) {
            const { modelIdMap: v } = y;
            await m.set(g, v);
          }
          await l.core.update(!0);
        }
      );
    }
    h.addEventListener(
      "sectionclick",
      async (f) => {
        const { values: w } = f.detail, _ = [];
        for (const g of w) {
          const { data: y } = g;
          if (!y)
            continue;
          const { modelIdMap: v } = y;
          _.push(
            d.highlightByID("select", v, !0, a.zoom)
          );
        }
        await Promise.all(_);
      }
    );
  };
  return c.html`
    <bim-chart ${c.ref(p)} type=${t ?? "bar"}>
      <bim-label slot="missing-data" icon="ph:warning-fill" style="--bim-icon--c: gold;">${u}</bim-label>
      ${s ? c.html`<bim-chart-legend slot="labels"></bim-chart-legend>` : ""}
    </bim-chart>
  `;
}, Ze = (r) => c.Component.create(
  Ke,
  r
), Qe = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  attributesChart: Ze
}, Symbol.toStringTag, { value: "Module" })), hi = {
  ...Ae,
  ...Ve,
  ...We,
  ...Ye,
  ...Qe
};
export {
  ci as Manager,
  lt as SheetBoard,
  P as ViewCube,
  Q as World,
  N as World2D,
  H as baseTopicTagStyle,
  di as buttons,
  hi as charts,
  G as createAuthorTag,
  E as defaultTopicStyles,
  pi as forms,
  ui as sections,
  mi as tables,
  gt as topicFormTemplate
};
